<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><title>SWISS-MODEL | Workspace | Model Results | kcedjm</title><style>
			.sequence { font-size: 12px; font-family:"Courier New", Courier, monospace;}
			h2 {font-size: 1.2em; font-family:Arial; margin:20 0 0 0}
			h3 {font-size: 1.0em; font-family:Arial; margin:0 0 0 0}
			
			.borderedTbl>tbody>tr>td, .borderedTbl th{ 
        text-align:center;
  			border-top:1px solid #DDD;
  			border-left:1px solid #DDD;
      }
      .borderedTbl{
        border-bottom:1px solid #DDD;
        border-right:1px solid #DDD; 
      }
      td { 
        -webkit-hyphens: auto;
           -moz-hyphens: auto;
                hyphens: auto; 
              }

			body {
			width:100%; 
			height:100%; 
			background-color: #E7E8EA;
			margin: 0;
			padding: 0;
			word-wrap: break-word;
			font-family: "Helvetica Neue",Helvetica,Arial,sans-serif;
			font-size: 10pt;
			}
			#content{
			padding:20px;
			width:780px;
			margin-left:auto;
			margin-right:auto;
			background-color:#FFF;}
			
			#sib_footer {
			background-color: #333333;
			color: #F7F8FA;
			font-family: Verdana;
			font-size: 10px;
			font-weight: normal;
			height: 20px;
			letter-spacing: 1px;
			
			width:760px;
      margin-left:auto;
      margin-right:auto;
      padding:0px 20px 0px 20px
			}
			
			#sib_footer a {
  			color: #FFFFFF;
  			font-family: Verdana;
  			font-size: 10px;
  			font-weight: normal;
  			letter-spacing: 1px;
			}
			a, a:link, a:visited {
			color: #D6161D;
			text-decoration: none;
			} 	

			.alTbl{ font-family:'Courier New', Courier, monospace; }
	   img{ border:0; }
			
	@media print{
	
	#content{
			padding:20px;
			background-color:#FFF;
		}
	body{ 
	    width:100%;
	    height:100%; 
	    background-color: #FFF;
			padding: 0;
		}	
	#sib_footer{
		display:none;
	}
	}
      
    </style></head><body><div id="content"><table width="100%" border="0"><tbody><tr><td valign="middle"><a title="SIB Swiss Institute of Bioinformatics" href="https://www.sib.swiss/"><img src="SWISS-MODEL-Q93VI0-2018_files/SIB_logo_20_sm.png"></a></td><td valign="middle"><a title="Biozentrum, University of Basel" href="https://www.biozentrum.unibas.ch/"><img src="SWISS-MODEL-Q93VI0-2018_files/biozentrum_en.png"></a></td><td valign="middle"><div style="font-size:1.9em;">SWISS-MODEL Homology Modelling Report</div></td></tr></tbody></table><div style="clear:both"></div><h2>Model Building Report</h2><p>
This document lists the results for the homology modelling project "Q93VI0" submitted to SWISS-MODEL workspace
on Feb. 28, 2018, 12:17 p.m..The submitted primary amino acid sequence is given in Table T1.</p><p>If you use any results in your research, please cite the relevant publications:</p><ul><li>Biasini, M., Bienert, S., Waterhouse, A., Arnold, K., Studer, G.,
    Schmidt, T., Kiefer, F., Cassarino, T.G., Bertoni, M., Bordoli, L.,
    Schwede, T. SWISS-MODEL: modelling protein tertiary and quaternary
    structure using evolutionary information. Nucleic Acids Res. 42, W252-W258
    (2014).
    <a href="https://www.ncbi.nlm.nih.gov/pubmed/24782522" target="_blank"><img src="SWISS-MODEL-Q93VI0-2018_files/pubmed.jpg" width="17" height="16" align="absmiddle"></a><a href="https://dx.doi.org/10.1093/nar/gku340" target="_blank"><img src="SWISS-MODEL-Q93VI0-2018_files/doi.jpg" style="padding-left:2px" width="30" height="16" align="absmiddle"></a></li><li><a name="promodII"></a>Guex, N., Peitsch, M.C., Schwede, T. Automated
          comparative protein structure modeling with SWISS-MODEL and
          Swiss-PdbViewer: A historical perspective. Electrophoresis 30,
          S162-S173 (2009).
    <a href="https://www.ncbi.nlm.nih.gov/pubmed/19517507" target="_blank"><img src="SWISS-MODEL-Q93VI0-2018_files/pubmed.jpg" width="17" height="16" align="absmiddle"></a><a href="https://dx.doi.org/10.1002/elps.200900140" target="_blank"><img src="SWISS-MODEL-Q93VI0-2018_files/doi.jpg" style="padding-left:2px" width="30" height="16" align="absmiddle"></a></li><li>Bienert, S., Waterhouse, A., de Beer, T.A., Tauriello, G., Studer,
    G., Bordoli, L., Schwede, T. The SWISS-MODEL Repository - new features and
    functionality. Nucleic Acids Res. 45, D313-D319 (2017).
    <a href="https://www.ncbi.nlm.nih.gov/pubmed/27899672" target="_blank"><img src="SWISS-MODEL-Q93VI0-2018_files/pubmed.jpg" width="17" height="16" align="absmiddle"></a><a href="https://dx.doi.org/10.1093/nar/gkw1132" target="_blank"><img src="SWISS-MODEL-Q93VI0-2018_files/doi.jpg" style="padding-left:2px" width="30" height="16" align="absmiddle"></a></li><li><a name="qmean"></a>Benkert, P., Biasini, M., Schwede, T. Toward the
        estimation of the absolute quality of individual protein structure
        models. Bioinformatics 27,
        343-350 (2011).
    <a href="https://www.ncbi.nlm.nih.gov/pubmed/21134891" target="_blank"><img src="SWISS-MODEL-Q93VI0-2018_files/pubmed.jpg" width="17" height="16" align="absmiddle"></a><a href="https://dx.doi.org/10.1093/bioinformatics/btq662" target="_blank"><img src="SWISS-MODEL-Q93VI0-2018_files/doi.jpg" style="padding-left:2px" width="30" height="16" align="absmiddle"></a></li><li><a name="qsqe"></a>Bertoni, M., Kiefer, F., Biasini, M., Bordoli, L.,
        Schwede, T. Modeling protein quaternary structure of homo- and
        hetero-oligomers beyond binary interactions by homology. Scientific
        Reports 7 (2017).
    <a href="https://www.ncbi.nlm.nih.gov/pubmed/28874689" target="_blank"><img src="SWISS-MODEL-Q93VI0-2018_files/pubmed.jpg" width="17" height="16" align="absmiddle"></a><a href="https://dx.doi.org/10.1038/s41598-017-09654-8" target="_blank"><img src="SWISS-MODEL-Q93VI0-2018_files/doi.jpg" style="padding-left:2px" width="30" height="16" align="absmiddle"></a></li></ul><p></p><h2>Results</h2><p>
The SWISS-MODEL template library (SMTL version 2018-02-21, PDB release 2018-02-16) was searched with 
BLAST (<a href="#blastp">Camacho et al.</a>) and HHBlits (<a href="#hhblits">Remmert et al.</a>)
 

 for evolutionary related structures matching the target sequence in 
Table T1. For details on the template search, see Materials and Methods.
 Overall 55 templates were found (Table T2).</p><h2>Models</h2><p>The following model was built (see Materials and Methods "Model Building"):</p><table class="borderedTbl" width="100%" cellspacing="1" cellpadding="5" border="0"><thead><tr><th><h3>Model #01</h3></th><th>File</th><th>Built with</th><th>Oligo-State</th><th>Ligands</th><th title="Global Model Quality Estimate">GMQE</th><th>QMEAN</th></tr></thead><tbody><tr><td><a href="https://swissmodel.expasy.org/interactive/kcedjm/models/01.png" target="_blank"><img src="SWISS-MODEL-Q93VI0-2018_files/01.png" height="60"></a></td><td><a href="https://swissmodel.expasy.org/interactive/kcedjm/models/01.pdb" target="_blank">PDB</a></td><td>
ProMod3 Version 1.1.0.</td><td>homo-dimer 
             
             (matching prediction)
             
          </td><td><div style="max-width:350px;">None</div></td><td>0.16</td><td>-4.01</td></tr></tbody></table><table width="100%"><tbody><tr><td><table cellspacing="2" cellpadding="0"><tbody><tr><td width="80">QMEAN</td><td>-4.01</td><td><img src="SWISS-MODEL-Q93VI0-2018_files/qmean4_compact.png" width="100" height="16"></td></tr><tr><td>CÎ²</td><td>-2.08</td><td><img src="SWISS-MODEL-Q93VI0-2018_files/cbeta_compact.png" width="100" height="16"></td></tr><tr><td>All Atom</td><td>-1.76</td><td><img src="SWISS-MODEL-Q93VI0-2018_files/all_atom_compact.png" width="100" height="16"></td></tr><tr><td>Solvation</td><td>-0.04</td><td><img src="SWISS-MODEL-Q93VI0-2018_files/solvation_compact.png" width="100" height="16"></td></tr><tr><td>Torsion</td><td>-3.34</td><td><img src="SWISS-MODEL-Q93VI0-2018_files/torsion_compact.png" width="100" height="16"></td></tr></tbody></table></td><td><span class="pull-left" style="margin-right:4px;"><a href="https://swissmodel.expasy.org/interactive/kcedjm/models/01/qmean/local_quality_estimate_A.png" target="_blank"><img src="SWISS-MODEL-Q93VI0-2018_files/local_quality_estimate_A.png" width="180" height="120"></a><span class="pull-left" style="margin-right:4px;"><a href="https://swissmodel.expasy.org/interactive/kcedjm/models/01/qmean/local_quality_estimate_B.png" target="_blank"><img src="SWISS-MODEL-Q93VI0-2018_files/local_quality_estimate_B.png" width="180" height="120"></a></span></span></td><td><a href="https://swissmodel.expasy.org/interactive/kcedjm/models/01/qmean/reference_set.png" target="_blank"><img src="SWISS-MODEL-Q93VI0-2018_files/reference_set.png" width="180" height="120"></a></td></tr></tbody></table><table class="borderedTbl"><thead><tr><th>Template</th><th>Seq Identity</th><th>Oligo-state</th><th>QSQE</th><th>Found by</th><th>Method</th><th>Resolution</th><th>Seq Similarity</th><th>Range</th><th>Coverage</th><th>Description</th></tr></thead><tbody><tr><td><a href="https://swissmodel.expasy.org/interactive/kcedjm/templates/23cc98fdc593e30a40fc5384e2fc7072c00c1727.pdb" target="_blank">2lcx.1.A</a></td><td>36.67</td><td>homo-dimer</td><td>0.00</td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.36</td><td>15 - 45</td><td>0.41</td><td>Receptor tyrosine-protein kinase erbB-4</td></tr></tbody></table> 
  	

    
  The template contained no ligands.
  

<br><pre>Target    MAAEFDGKIESKGLNPG-LIVLLVIGGLLLTFLVGNFI-LYTYAQKNLPPRKKKPVSKKKMKKEKMKQGVQVPGE<br>2lcx.1.A  --------------ARTPLIAAGVIGGLFILVIVGLTFAVYV-RRKS----------------------------<br><br></pre><br><hr><br><h2>Materials and Methods</h2><h2>Template Search</h2><p>Template
 search with BLAST and HHBlits
 has been performed against the SWISS-MODEL template library (SMTL, last
 update: 2018-02-21, last included PDB release: 2018-02-16).</p><p>
The target sequence was searched with BLAST against the primary amino acid sequence contained in the SMTL.



</p><p>
An initial HHblits profile has been built using the procedure outlined 
in (Remmert et al.), followed by 1 iteration of HHblits against NR20. 
The obtained profile has then be searched against all profiles of the 
SMTL. A total of 59 templates were found.</p><h2>Template Selection</h2><p>For each identified template, the template's quality has been predicted from features of the target-template alignment.
 The templates with the highest quality have then been selected for model building.</p><h2>Model Building</h2><p>
Models are built based on the target-template alignment using ProMod3. 
Coordinates which are conserved between the target and the template are 
copied from the template to the model. Insertions and deletions are 
remodelled using a fragment library. Side chains are then rebuilt. 
Finally, the geometry of the resulting model is regularized by using a 
force field. In case loop modelling with ProMod3 fails, an alternative 
model is built with PROMOD-II (<a href="#promodII">Guex et al.</a>).
</p><h2>Model Quality Estimation</h2><p>
The global and per-residue model quality has been assessed using the QMEAN scoring function (<a href="#qmean">Benkert et al.</a>) . For improved performance, weights of the individual QMEAN terms have been trained specifically for SWISS-MODEL. 
</p><h2>Ligand Modelling</h2><p>
Ligands present in the template structure are transferred by homology to
 the model when the following criteria are met: (a) The ligands are 
annotated as biologically relevant in the template library, (b) the 
ligand is in contact with the model, (c) the ligand is not clashing with
 the protein, (d) the residues in contact with the ligand are conserved 
between the target and the template. If any of these four criteria is 
not satisfied, a certain ligand will not be included in the model. The 
model summary includes information on why and which ligand has not been 
included.</p><h2>Oligomeric State Conservation</h2><p>
The quaternary structure annotation of the template is used to model the target sequence in its oligomeric form. The method (<a href="#qsqe">Bertoni et al.</a>)
 is based on a supervised machine learning algorithm, Support Vector 
Machines (SVM), which combines interface conservation, structural 
clustering, and other template features to provide a quaternary 
structure quality estimate (QSQE). The QSQE score is a number between 0 
and 1, reflecting the expected accuracy of the interchain contacts for a
 model built based a given alignment and template. Higher numbers 
indicate higher reliability. This complements the GMQE score which 
estimates the accuracy of the tertiary structure of the resulting model.</p><h2>References</h2><p></p><ul><li><a name="blastp"></a><b>BLAST</b><br>
        Camacho, C., Coulouris, G., Avagyan, V., Ma, N., Papadopoulos, J.,
        Bealer, K., Madden, T.L. BLAST+: architecture and applications. BMC
        Bioinformatics 10, 421-430 (2009).
        <a target="_blank" href="https://www.ncbi.nlm.nih.gov/pubmed/20003500"><img src="SWISS-MODEL-Q93VI0-2018_files/pubmed.jpg" width="17" height="16" align="absmiddle"></a><a href="https://dx.doi.org/10.1186/1471-2105-10-421" target="_blank"><img src="SWISS-MODEL-Q93VI0-2018_files/doi.jpg" style="padding-left:2px" width="30" height="16" align="absmiddle"></a></li><li><a name="hhblits"></a><b>HHblits</b><br>
        Remmert, M., Biegert, A., Hauser, A., SÃ¶ding, J. HHblits:
        lightning-fast iterative protein sequence searching by HMM-HMM
        alignment. Nat Methods 9, 173-175 (2012).
        <a target="_blank" href="https://www.ncbi.nlm.nih.gov/pubmed/22198341"><img src="SWISS-MODEL-Q93VI0-2018_files/pubmed.jpg" width="17" height="16" align="absmiddle"></a><a href="https://dx.doi.org/10.1038/nmeth.1818" target="_blank"><img src="SWISS-MODEL-Q93VI0-2018_files/doi.jpg" style="padding-left:2px" width="30" height="16" align="absmiddle"></a></li></ul><h2>Table T1:</h2><p>Primary amino acid sequence for which templates were searched and models were built.</p><div class="sequence">MAAEFDGKIESKGLNPGLIVLLVIGGLLLTFLVGNFILYTYAQKNLPPRKKKPVSKKKMKKEKMKQGVQVPGE<br></div><h2>Table T2:</h2><table class="borderedTbl"><thead><tr><th>Template</th><th>Seq Identity</th><th>Oligo-state</th><th>QSQE</th><th>Found by</th><th>Method</th><th>Resolution</th><th>Seq Similarity</th><th>Coverage</th><th>Description</th></tr></thead><tbody><tr><td>2lcx.1.A</td><td>36.67</td><td>homo-dimer</td><td></td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.36</td><td>0.41</td><td>Receptor tyrosine-protein kinase erbB-4</td></tr><tr><td>2lcx.1.B</td><td>36.67</td><td>homo-dimer</td><td></td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.36</td><td>0.41</td><td>Receptor tyrosine-protein kinase erbB-4</td></tr><tr><td>2l2t.1.A</td><td>36.67</td><td>homo-dimer</td><td>0.08</td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.36</td><td>0.41</td><td>Receptor tyrosine-protein kinase erbB-4</td></tr><tr><td>2l2t.1.B</td><td>36.67</td><td>homo-dimer</td><td>0.08</td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.36</td><td>0.41</td><td>Receptor tyrosine-protein kinase erbB-4</td></tr><tr><td>4wo1.2.B</td><td>27.59</td><td>homo-dimer</td><td>0.16</td><td>HHblits</td><td>X-ray</td><td>2.14Ã</td><td>0.36</td><td>0.40</td><td>TYRO protein tyrosine kinase-binding protein</td></tr><tr><td>2l34.1.B</td><td>27.59</td><td>homo-dimer</td><td>0.15</td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.36</td><td>0.40</td><td>TYRO protein tyrosine kinase-binding protein</td></tr><tr><td>4wo1.2.A</td><td>27.59</td><td>homo-dimer</td><td>0.16</td><td>HHblits</td><td>X-ray</td><td>2.14Ã</td><td>0.36</td><td>0.40</td><td>TYRO protein tyrosine kinase-binding protein</td></tr><tr><td>2l34.1.A</td><td>27.59</td><td>homo-dimer</td><td>0.15</td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.36</td><td>0.40</td><td>TYRO protein tyrosine kinase-binding protein</td></tr><tr><td>4wol.1.C</td><td>27.59</td><td>homo-trimer</td><td>0.04</td><td>HHblits</td><td>X-ray</td><td>1.77Ã</td><td>0.36</td><td>0.40</td><td>TYRO protein tyrosine kinase-binding protein</td></tr><tr><td>4wol.1.B</td><td>27.59</td><td>homo-trimer</td><td>0.04</td><td>HHblits</td><td>X-ray</td><td>1.77Ã</td><td>0.36</td><td>0.40</td><td>TYRO protein tyrosine kinase-binding protein</td></tr><tr><td>2l35.1.B</td><td>27.59</td><td>monomer</td><td></td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.36</td><td>0.40</td><td>TYRO protein tyrosine kinase-binding protein</td></tr><tr><td>2n9y.1.A</td><td>22.58</td><td>monomer</td><td></td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.31</td><td>0.42</td><td>Integrin alpha-IIb</td></tr><tr><td>2mfr.1.A</td><td>27.59</td><td>monomer</td><td></td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.34</td><td>0.40</td><td>Insulin receptor</td></tr><tr><td>2bg9.1.C</td><td>20.69</td><td>monomer</td><td></td><td>HHblits</td><td>EM</td><td>4.00Ã</td><td>0.32</td><td>0.40</td><td>ACETYLCHOLINE RECEPTOR PROTEIN, DELTA CHAIN</td></tr><tr><td>2knc.1.A</td><td>20.69</td><td>monomer</td><td></td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.31</td><td>0.40</td><td>Integrin alpha-IIb</td></tr><tr><td>2k9j.1.A</td><td>21.43</td><td>monomer</td><td></td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.31</td><td>0.38</td><td>Integrin alpha-IIb light chain</td></tr><tr><td>2k1a.1.A</td><td>21.43</td><td>monomer</td><td></td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.31</td><td>0.38</td><td>Integrin alpha-IIb</td></tr><tr><td>5lv6.1.B</td><td>29.63</td><td>homo-dimer</td><td></td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.33</td><td>0.37</td><td>Epidermal growth factor receptor</td></tr><tr><td>2m0b.1.A</td><td>29.63</td><td>homo-dimer</td><td></td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.33</td><td>0.37</td><td>Epidermal growth factor receptor</td></tr><tr><td>2m0b.1.B</td><td>29.63</td><td>homo-dimer</td><td></td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.33</td><td>0.37</td><td>Epidermal growth factor receptor</td></tr><tr><td>2ks1.1.B</td><td>29.63</td><td>monomer</td><td></td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.33</td><td>0.37</td><td>Epidermal growth factor receptor</td></tr><tr><td>2l35.1.A</td><td>32.00</td><td>monomer</td><td></td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.38</td><td>0.34</td><td>DAP12-NKG2C_TM</td></tr><tr><td>2m3e.1.A</td><td>34.62</td><td>monomer</td><td></td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.32</td><td>0.36</td><td>Integrin alpha-L</td></tr><tr><td>2n2a.1.A</td><td>25.93</td><td>homo-dimer</td><td>0.12</td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.28</td><td>0.37</td><td>Receptor tyrosine-protein kinase erbB-2</td></tr><tr><td>2n2a.1.B</td><td>25.93</td><td>homo-dimer</td><td>0.12</td><td>HHblits</td><td>NMR</td><td>NA</td><td>0.28</td><td>0.37</td><td>Receptor tyrosine-protein kinase erbB-2</td></tr></tbody></table><br>The
 table above shows the top 25 filtered templates. A further 30 templates
 were found which were considered to be less suitable for modelling than
 the filtered list.<br>
3a0b.1.I, 1s5l.1.I, 1xme.1.C, 3kzi.1.I, 1ehk.1.C, 5v2c.1.I, 2l91.1.A, 
5xnl.1.N, 2lx0.1.A, 2n5s.1.A, 5djq.1.D, 4il6.1.I, 2m20.1.A, 2rn0.1.A, 
2l8s.1.A, 2rmz.1.A, 5gti.1.I, 2lzl.1.A, 2lzl.1.B, 5h1s.1.T, 2k9j.1.B, 
3jcu.1.J, 5li0.1.j, 4fby.1.I, 4rvy.1.Q, 5e79.1.I, 2knc.1.B, 3s39.1.C, 
3a0h.1.I, 4yuu.1.I
 


	</div><div id="sib_footer"><div id="sib_footer_content"><a href="https://www.sib.swiss/">Swiss Institute of Bioinformatics</a><div id="sib_footer_right" style="float:right"></div></div></div>
</body></html>