{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"images/JHI_STRAP_Web.png\" style=\"width: 150px; float: right;\">\n",
    "# XX - Runthrough"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is essentially a crib sheet for the tutor.\n",
    "\n",
    "* **Reinforce the value of reproducible research, and literate programming (embedding runnable code and analysis in the same document)**\n",
    "* **Ask the learners to write a short introduction about the goals of the report**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 00 - Python Imports\n",
    "\n",
    "You'll need the following imports:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "# Show graphics inline\n",
    "%pylab inline\n",
    "\n",
    "# PSL\n",
    "import os\n",
    "\n",
    "# Import Pandas and Seaborn\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "# Biopython\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Blast import NCBIWWW\n",
    "from Bio.Blast.Applications import NcbiblastxCommandline, NcbiblastnCommandline, \\\n",
    "                                   NcbitblastnCommandline, NcbitblastxCommandline\n",
    "    \n",
    "# bioservices\n",
    "from bioservices import UniProt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Also to make an output directory - creation at the command-line is OK."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Create output directory if needed\n",
    "os.makedirs(\"output\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 01 - Mapping/searching for CDS on a genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 01a - Load in the sequences\n",
    "\n",
    "* **The downstream analyses require us to have the sequences available**\n",
    "* **Ask learners to write a little summary of sequence source information**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert-success\">\n",
    "The nucleotide sequences for wildtype (`wildtype`) and engineered (`engineered`) lipases as provided by the Product Development Unit are loaded.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">wildtype lipase protein from Proteus mirabilis\n",
      "ATGGGCAGCAGCCACCACCACCACCACCACAGCAGCGGCCTGGTGCCCAGGGGCAGCCAC\n",
      "ATGAGCACCAAGTACCCCATCGTGCTGGTGCACGGCCTGGCCGGCTTCAACGAGATCGTG\n",
      "GGCTTCCCCTACTTCTACGGCATCGCCGACGCCCTGAGGCAGGACGGCCACCAGGTGTTC\n",
      "ACCGCCAGCCTGAGCGCCTTCAACAGCAACGAGGTGAGGGGCAAGCAGCTGTGGCAGTTC\n",
      "GTGCAGACCCTGCTGCAGGAGACCCAGGCCAAGAAGGTGAACTTCATCGGCCACAGCCAG\n",
      "GGCCCCCTGGCCTGCAGGTACGTGGCCGCCAACTACCCCGACAGCGTGGCCAGCGTGACC\n",
      "AGCATCAACGGCGTGAACCACGGCAGCGAGATCGCCGACCTGTACAGGAGGATCATGAGG\n",
      "AAGGACAGCATCCCCGAGTACATCGTGGAGAAGGTGCTGAACGCCTTCGGCACCATCATC\n",
      "AGCACCTTCAGCGGCCACAGGGGCGACCCCCAGGACGCCATCGCCGCCCTGGAGAGCCTG\n",
      "ACCACCGAGCAGGTGACCGAGTTCAACAACAAGTACCCCCAGGCCCTGCCCAAGACCCCC\n",
      "GGCGGCGAGGGCGACGAGATCGTGAACGGCGTGCACTACTACTGCTTCGGCAGCTACATC\n",
      "CAGGGCCTGATCGCCGGCGAGAAGGGCAACCTGCTGGACCCCACCCACGCCGCCATGAGG\n",
      "GTGCTGAACACCTTCTTCACCGAGAAGCAGAACGACGGCCTGGTGGGCAGGAGCAGCATG\n",
      "AGGCTGGGCAAGCTGATCAAGGACGACTACGCCCAGGACCACATCGACATGGTGAACCAG\n",
      "GTGGCCGGCCTGGTGGGCTACAACGAGGACATCGTGGCCATCTACACCCAGCACGCCAAG\n",
      "TACCTGGCCAGCAAGCAGCTG\n",
      "\n",
      ">engineered lipase protein from Proteus mirabilis\n",
      "ATGGGCAGCAGCCACCACCACCACCACCACAGCAGCGGCCTGGTGCCCAGGGGCAGCCAC\n",
      "ATGAGCACCAAGTACCCCATCGTGCTGGTGCACGGCCTGGCCGGCTTCAGCGAGATCGTG\n",
      "GGCTTCCCCTACTTCTACGGCATCGCCGACGCCCTGACCCAGGACGGCCACCAGGTGTTC\n",
      "ACCGCCAGCCTGAGCGCCTTCAACAGCAACGAGGTGAGGGGCAAGCAGCTGTGGCAGTTC\n",
      "GTGCAGACCATCCTGCAGGAGACCCAGACCAAGAAGGTGAACTTCATCGGCCACAGCCAG\n",
      "GGCCCCCTGGCCTGCAGGTACGTGGCCGCCAACTACCCCGACAGCGTGGCCAGCGTGACC\n",
      "AGCATCAACGGCGTGAACCACGGCAGCGAGATCGCCGACCTGTACAGGAGGATCATCAGG\n",
      "AAGGACAGCATCCCCGAGTACATCGTGGAGAAGGTGCTGAACGCCTTCGGCACCATCATC\n",
      "AGCACCTTCAGCGGCCACAGGGGCGACCCCCAGGACGCCATCGCCGCCCTGGAGAGCCTG\n",
      "ACCACCGAGCAGGTGACCGAGTTCAACAACAAGTACCCCCAGGCCCTGCCCAAGACCCCC\n",
      "TGCGGCGAGGGCGACGAGATCGTGAACGGCGTGCACTACTACTGCTTCGGCAGCTACATC\n",
      "CAGGAGCTGATCGCCGGCGAGAACGGCAACCTGCTGGACCCCACCCACGCCGCCATGAGG\n",
      "GTGCTGAACACCCTGTTCACCGAGAAGCAGAACGACGGCCTGGTGGGCAGGTGCAGCATG\n",
      "AGGCTGGGCAAGCTGATCAAGGACGACTACGCCCAGGACCACTTCGACATGGTGAACCAG\n",
      "GTGGCCGGCCTGGTGAGCTACAACGAGAACATCGTGGCCATCTACACCCTGCACGCCAAG\n",
      "TACCTGGCCAGCAAGCAGCTG\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Read the wildtype and engineered sequences in from file\n",
    "wildtype = SeqIO.read(os.path.join('data', 'wildtype_nt.fasta'), 'fasta')\n",
    "engineered = SeqIO.read(os.path.join('data', 'engineered_nt.fasta'), 'fasta')\n",
    "\n",
    "# Show the sequences\n",
    "print(wildtype.format('fasta'))\n",
    "print(engineered.format('fasta'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lengths - WT: 921nt, ENG: 921nt\n"
     ]
    }
   ],
   "source": [
    "# Summarise sequence properties\n",
    "# Lengths\n",
    "print(\"lengths - WT: {0}nt, ENG: {1}nt\".format(len(wildtype), len(engineered)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Count of differing bases: 18\n"
     ]
    }
   ],
   "source": [
    "# Count and report differences between the sequences (OPTIONAL)\n",
    "diffcount = 0\n",
    "for pos in range(len(wildtype)):\n",
    "    if engineered[pos] != wildtype[pos]:\n",
    "        diffcount += 1\n",
    "print(\"Count of differing bases: {0}\".format(diffcount))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert-success\">\n",
    "The wildtype and engineered sequences are each 921bp in length, with 18 SNPs that distinguish the engineered form.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Convert nucleotide sequences to protein (use table 11)**\n",
    "* **Note that sequence IDs can't have spaces**\n",
    "* **Ask learners to write sequences to file**\n",
    "* **Ask learners to write short summary of conversion process**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert-success\">\n",
    "The nucleotide sequences are translated into predicted gene products using the `Biopython` `translate()` method, with translation table 11 (bacteria, archaea, plant plastid). \n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">translated_wildtype wildtype lipase protein from Proteus mirabilis\n",
      "MGSSHHHHHHSSGLVPRGSHMSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVF\n",
      "TASLSAFNSNEVRGKQLWQFVQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVT\n",
      "SINGVNHGSEIADLYRRIMRKDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESL\n",
      "TTEQVTEFNNKYPQALPKTPGGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMR\n",
      "VLNTFFTEKQNDGLVGRSSMRLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAK\n",
      "YLASKQL\n",
      "\n",
      ">translated_engineered engineered lipase protein from Proteus mirabilis\n",
      "MGSSHHHHHHSSGLVPRGSHMSTKYPIVLVHGLAGFSEIVGFPYFYGIADALTQDGHQVF\n",
      "TASLSAFNSNEVRGKQLWQFVQTILQETQTKKVNFIGHSQGPLACRYVAANYPDSVASVT\n",
      "SINGVNHGSEIADLYRRIIRKDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESL\n",
      "TTEQVTEFNNKYPQALPKTPCGEGDEIVNGVHYYCFGSYIQELIAGENGNLLDPTHAAMR\n",
      "VLNTLFTEKQNDGLVGRCSMRLGKLIKDDYAQDHFDMVNQVAGLVSYNENIVAIYTLHAK\n",
      "YLASKQL\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Translate input nucleotide sequences\n",
    "wildtype_aa = SeqRecord(id=\"translated_\" + wildtype.id,\n",
    "                        description=wildtype.description,\n",
    "                        seq = wildtype.seq.translate(table=11))\n",
    "engineered_aa = SeqRecord(id=\"translated_\" + engineered.id,\n",
    "                        description=engineered.description,\n",
    "                        seq = engineered.seq.translate(table=11))\n",
    "\n",
    "# Show the sequences\n",
    "print(wildtype_aa.format('fasta'))\n",
    "print(engineered_aa.format('fasta'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lengths - WT: 307aa, ENG: 307aa\n"
     ]
    }
   ],
   "source": [
    "# Summarise sequence properties\n",
    "# Lengths\n",
    "print(\"lengths - WT: {0}aa, ENG: {1}aa\".format(len(wildtype_aa), len(engineered_aa)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Count of differing residues: 14\n"
     ]
    }
   ],
   "source": [
    "# Count and report differences between the sequences (OPTIONAL)\n",
    "diffcount = 0\n",
    "for pos in range(len(wildtype_aa)):\n",
    "    if engineered_aa[pos] != wildtype_aa[pos]:\n",
    "        diffcount += 1\n",
    "print(\"Count of differing residues: {0}\".format(diffcount))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Write sequences to file\n",
    "SeqIO.write(wildtype_aa, os.path.join(\"output\", \"wildtype_aa.fasta\"), \"fasta\")\n",
    "SeqIO.write(engineered_aa, os.path.join(\"output\", \"engineered_aa.fasta\"), \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert-success\">\n",
    "This produces conceptual translations of 307aa in length, with 14 residue substitutions between wild-type and engineered forms. These sequences were written to `output/wildtype_aa.fasta` and `output/engineered_aa.fasta`.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 01b - Identifying a gene model\n",
    "\n",
    "* **The wildtype nt sequence will be `BLAST`ed against the genome to identify a likely location; this will be visualised against the `.gbff` file for the reference.**\n",
    "* **Ask the learners to search for the wildtype sequence on the genome.**\n",
    "* **Allow the learners to use the command-line or notebook to conduct the search**\n",
    "* **In either case, the first step is database creation, which requires dropping to the command-line**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert-success\">\n",
    "The following local `BLAST` databases of the <i>P. mirabilis</i> reference genome are created in the `output` directory, with the names:\n",
    "<ul>\n",
    "    <li> The reference genome sequence (nucleotide) - `ref_genome`\n",
    "    <li> The reference annotated CDS (nucleotide) - `ref_cds`\n",
    "    <li> The reference annotated proteins (protein) - `ref_protein`\n",
    "</ul>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 03/13/2017 15:40:30\n",
      "New DB name:   /Users/lpritc/Documents/Development/Teaching/Teaching-IBioIC-Intro-to-Bioinformatics/03-lipases/output/ref_genome\n",
      "New DB title:  ref_genome\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Users/lpritc/Documents/Development/Teaching/Teaching-IBioIC-Intro-to-Bioinformatics/03-lipases/output/ref_genome\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 2 sequences in 0.053622 seconds.\n",
      "\n",
      "\n",
      "Building a new DB, current time: 03/13/2017 15:40:30\n",
      "New DB name:   /Users/lpritc/Documents/Development/Teaching/Teaching-IBioIC-Intro-to-Bioinformatics/03-lipases/output/ref_cds\n",
      "New DB title:  ref_cds\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Users/lpritc/Documents/Development/Teaching/Teaching-IBioIC-Intro-to-Bioinformatics/03-lipases/output/ref_cds\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 3740 sequences in 0.195287 seconds.\n",
      "\n",
      "\n",
      "Building a new DB, current time: 03/13/2017 15:40:31\n",
      "New DB name:   /Users/lpritc/Documents/Development/Teaching/Teaching-IBioIC-Intro-to-Bioinformatics/03-lipases/output/ref_protein\n",
      "New DB title:  ref_protein\n",
      "Sequence type: Protein\n",
      "Deleted existing Protein BLAST database named /Users/lpritc/Documents/Development/Teaching/Teaching-IBioIC-Intro-to-Bioinformatics/03-lipases/output/ref_protein\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 3662 sequences in 0.129818 seconds.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# Drop out to bash to create BLAST databases - can do this at the command line\n",
    "\n",
    "# reference genome\n",
    "makeblastdb -in data/GCA_000069965.1_ASM6996v1_genomic.fna \\\n",
    "            -dbtype nucl \\\n",
    "            -title ref_genome \\\n",
    "            -out output/ref_genome\n",
    "            \n",
    "# reference CDS            \n",
    "makeblastdb -in data/GCA_000069965.1_ASM6996v1_cds_from_genomic.fna \\\n",
    "            -dbtype nucl \\\n",
    "            -title ref_cds \\\n",
    "            -out output/ref_cds\n",
    "            \n",
    "# reference protein\n",
    "makeblastdb -in data/GCA_000069965.1_ASM6996v1_protein.faa \\\n",
    "            -dbtype prot \\\n",
    "            -title ref_protein \\\n",
    "            -out output/ref_protein"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **However the BLAST search is run, note the utility of having an informative output filename**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert-success\">\n",
    "A `BLASTN` search using the wildtype sequence as query was carried out against the `ref_genome` database to identify a likely genomic location.\n",
    "</div>\n",
    "\n",
    "**Option 1: command-line search**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Command-line search\n",
    "blastn -query data/wildtype_nt.fasta \\\n",
    "       -db output/ref_genome \\\n",
    "       -outfmt 6 \\\n",
    "       -out output/wildtype_blastn_ref_genome.tab"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "**Option 2: notebook search**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STDOUT: \n",
      "STDERR: \n"
     ]
    }
   ],
   "source": [
    "# Create BLASTN command line\n",
    "cmd_blastn = NcbiblastnCommandline(query=os.path.join(\"data\", \"wildtype_nt.fasta\"),\n",
    "                                   db=os.path.join(\"output\", \"ref_genome\"),\n",
    "                                   outfmt=6,\n",
    "                                   out=os.path.join(\"output\", \"wildtype_blastn_ref_genome.tab\"))\n",
    "\n",
    "# Run BLASTN command\n",
    "stdout, stderr = cmd_blastn()\n",
    "print(\"STDOUT: %s\" % stdout)\n",
    "print(\"STDERR: %s\" % stderr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Check the output file. This search provides no matches!**\n",
    "* **Ask the learners to consider a different approach - guide them towards `TBLASTX` (translated query, translated subject)**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert-success\">\n",
    "As the `BLASTN` search produced no matches, a `TBLASTX` search was used, with the same input sequence and database.\n",
    "</div>\n",
    "\n",
    "**Option 1: command-line search**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Command-line search\n",
    "tblastx -query data/wildtype_nt.fasta \\\n",
    "        -db output/ref_genome \\\n",
    "        -outfmt 6 \\\n",
    "        -out output/wildtype_tblastx_ref_genome.tab"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Option 2: notebook search**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STDOUT: \n",
      "STDERR: \n"
     ]
    }
   ],
   "source": [
    "# Create TBLASTX command line\n",
    "cmd_tblastx = NcbitblastxCommandline(query=os.path.join(\"data\", \"wildtype_nt.fasta\"),\n",
    "                                    db=os.path.join(\"output\", \"ref_genome\"),\n",
    "                                    outfmt=6,\n",
    "                                    out=os.path.join(\"output\", \"wildtype_tblastx_ref_genome.tab\"))\n",
    "\n",
    "# Run BLASTN command\n",
    "stdout, stderr = cmd_tblastx()\n",
    "print(\"STDOUT: %s\" % stdout)\n",
    "print(\"STDERR: %s\" % stderr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Ask the learners to read in the output file using `pandas`, and present the matches with appropriate headers and a comment.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>subject</th>\n",
       "      <th>pc_identity</th>\n",
       "      <th>aln_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gaps_opened</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>subject_start</th>\n",
       "      <th>subject_end</th>\n",
       "      <th>e_value</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>AM942759.1</td>\n",
       "      <td>98.639</td>\n",
       "      <td>294</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>40</td>\n",
       "      <td>921</td>\n",
       "      <td>1063060</td>\n",
       "      <td>1063941</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>692.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>AM942759.1</td>\n",
       "      <td>70.769</td>\n",
       "      <td>130</td>\n",
       "      <td>38</td>\n",
       "      <td>0</td>\n",
       "      <td>752</td>\n",
       "      <td>363</td>\n",
       "      <td>1063772</td>\n",
       "      <td>1063383</td>\n",
       "      <td>2.820000e-90</td>\n",
       "      <td>197.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>AM942759.1</td>\n",
       "      <td>72.289</td>\n",
       "      <td>83</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>314</td>\n",
       "      <td>66</td>\n",
       "      <td>1063334</td>\n",
       "      <td>1063086</td>\n",
       "      <td>2.820000e-90</td>\n",
       "      <td>121.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>AM942759.1</td>\n",
       "      <td>80.769</td>\n",
       "      <td>26</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>920</td>\n",
       "      <td>843</td>\n",
       "      <td>1063940</td>\n",
       "      <td>1063863</td>\n",
       "      <td>2.820000e-90</td>\n",
       "      <td>48.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>AM942759.1</td>\n",
       "      <td>80.000</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>76</td>\n",
       "      <td>105</td>\n",
       "      <td>588332</td>\n",
       "      <td>588303</td>\n",
       "      <td>3.800000e-01</td>\n",
       "      <td>23.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>AM942759.1</td>\n",
       "      <td>35.294</td>\n",
       "      <td>34</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>202</td>\n",
       "      <td>303</td>\n",
       "      <td>588212</td>\n",
       "      <td>588111</td>\n",
       "      <td>3.800000e-01</td>\n",
       "      <td>23.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>AM942759.1</td>\n",
       "      <td>38.889</td>\n",
       "      <td>36</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>238</td>\n",
       "      <td>345</td>\n",
       "      <td>3623411</td>\n",
       "      <td>3623304</td>\n",
       "      <td>9.800000e-01</td>\n",
       "      <td>29.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>AM942759.1</td>\n",
       "      <td>43.478</td>\n",
       "      <td>23</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>787</td>\n",
       "      <td>855</td>\n",
       "      <td>1102334</td>\n",
       "      <td>1102402</td>\n",
       "      <td>9.800000e-01</td>\n",
       "      <td>29.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>AM942759.1</td>\n",
       "      <td>52.381</td>\n",
       "      <td>21</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>286</td>\n",
       "      <td>348</td>\n",
       "      <td>3298352</td>\n",
       "      <td>3298414</td>\n",
       "      <td>9.800000e-01</td>\n",
       "      <td>29.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>AM942759.1</td>\n",
       "      <td>66.667</td>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>530</td>\n",
       "      <td>495</td>\n",
       "      <td>3695049</td>\n",
       "      <td>3695084</td>\n",
       "      <td>3.500000e+00</td>\n",
       "      <td>27.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>AM942759.1</td>\n",
       "      <td>41.667</td>\n",
       "      <td>24</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>560</td>\n",
       "      <td>489</td>\n",
       "      <td>277868</td>\n",
       "      <td>277939</td>\n",
       "      <td>6.500000e+00</td>\n",
       "      <td>26.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>AM942759.1</td>\n",
       "      <td>60.000</td>\n",
       "      <td>15</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>690</td>\n",
       "      <td>734</td>\n",
       "      <td>1517270</td>\n",
       "      <td>1517314</td>\n",
       "      <td>8.600000e+00</td>\n",
       "      <td>26.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>AM942759.1</td>\n",
       "      <td>32.432</td>\n",
       "      <td>37</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>337</td>\n",
       "      <td>447</td>\n",
       "      <td>1135838</td>\n",
       "      <td>1135728</td>\n",
       "      <td>9.000000e+00</td>\n",
       "      <td>26.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       query     subject  pc_identity  aln_length  mismatches  gaps_opened  \\\n",
       "0   wildtype  AM942759.1       98.639         294           4            0   \n",
       "1   wildtype  AM942759.1       70.769         130          38            0   \n",
       "2   wildtype  AM942759.1       72.289          83          23            0   \n",
       "3   wildtype  AM942759.1       80.769          26           5            0   \n",
       "4   wildtype  AM942759.1       80.000          10           2            0   \n",
       "5   wildtype  AM942759.1       35.294          34          22            0   \n",
       "6   wildtype  AM942759.1       38.889          36          22            0   \n",
       "7   wildtype  AM942759.1       43.478          23          13            0   \n",
       "8   wildtype  AM942759.1       52.381          21          10            0   \n",
       "9   wildtype  AM942759.1       66.667          12           4            0   \n",
       "10  wildtype  AM942759.1       41.667          24          14            0   \n",
       "11  wildtype  AM942759.1       60.000          15           6            0   \n",
       "12  wildtype  AM942759.1       32.432          37          25            0   \n",
       "\n",
       "    query_start  query_end  subject_start  subject_end       e_value  bitscore  \n",
       "0            40        921        1063060      1063941  0.000000e+00     692.0  \n",
       "1           752        363        1063772      1063383  2.820000e-90     197.0  \n",
       "2           314         66        1063334      1063086  2.820000e-90     121.0  \n",
       "3           920        843        1063940      1063863  2.820000e-90      48.3  \n",
       "4            76        105         588332       588303  3.800000e-01      23.1  \n",
       "5           202        303         588212       588111  3.800000e-01      23.1  \n",
       "6           238        345        3623411      3623304  9.800000e-01      29.5  \n",
       "7           787        855        1102334      1102402  9.800000e-01      29.5  \n",
       "8           286        348        3298352      3298414  9.800000e-01      29.5  \n",
       "9           530        495        3695049      3695084  3.500000e+00      27.6  \n",
       "10          560        489         277868       277939  6.500000e+00      26.7  \n",
       "11          690        734        1517270      1517314  8.600000e+00      26.3  \n",
       "12          337        447        1135838      1135728  9.000000e+00      26.3  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define column headers\n",
    "headers = ['query', 'subject',\n",
    "           'pc_identity', 'aln_length', 'mismatches', 'gaps_opened',\n",
    "           'query_start', 'query_end', 'subject_start', 'subject_end',\n",
    "           'e_value', 'bitscore']\n",
    "\n",
    "# Read TBLASTX output and add headers\n",
    "results = pd.read_csv(os.path.join(\"output\", \"wildtype_tblastx_ref_genome.tab\"), sep=\"\\t\", header=None)\n",
    "results.columns = headers\n",
    "\n",
    "# Show results\n",
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert-success\">\n",
    "This query produced a number of candidate genomic region matches. The best match was to the region 1063060-1063941, covering 294 amino acids, at 98.6% identity. This match was taken to be the most likely genomic location for the wildtype sequence.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Ask the learners to search the annotated proteins from the reference, as for the genome search, and summarise the results in a similar way.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert-success\">\n",
    "A `BLASTX` search using the wildtype sequence as query was carried out against the `ref_protein` database to identify whether a corresponding protein has been annotated on the reference.\n",
    "</div>\n",
    "\n",
    "**Option 1: command-line search**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Command-line search\n",
    "blastx -query data/wildtype_nt.fasta \\\n",
    "       -db output/ref_protein \\\n",
    "       -outfmt 6 \\\n",
    "       -out output/wildtype_blastx_ref_protein.tab"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Option 2: notebook search**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STDOUT: \n",
      "STDERR: \n"
     ]
    }
   ],
   "source": [
    "# Create BLASTX command line\n",
    "cmd_blastx = NcbiblastxCommandline(query=os.path.join(\"data\", \"wildtype_nt.fasta\"),\n",
    "                                   db=os.path.join(\"output\", \"ref_protein\"),\n",
    "                                   outfmt=6,\n",
    "                                   out=os.path.join(\"output\", \"wildtype_blastx_ref_protein.tab\"))\n",
    "\n",
    "# Run BLASTN command\n",
    "stdout, stderr = cmd_blastx()\n",
    "print(\"STDOUT: %s\" % stdout)\n",
    "print(\"STDERR: %s\" % stderr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Remind the learners that they can reuse the headers they defined earlier, when presenting the results as a `pandas` dataframe**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>subject</th>\n",
       "      <th>pc_identity</th>\n",
       "      <th>aln_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gaps_opened</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>subject_start</th>\n",
       "      <th>subject_end</th>\n",
       "      <th>e_value</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>CAR42190.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>287</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61</td>\n",
       "      <td>921</td>\n",
       "      <td>1</td>\n",
       "      <td>287</td>\n",
       "      <td>0.00</td>\n",
       "      <td>595.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>CAR41317.1</td>\n",
       "      <td>29.412</td>\n",
       "      <td>102</td>\n",
       "      <td>62</td>\n",
       "      <td>3</td>\n",
       "      <td>76</td>\n",
       "      <td>369</td>\n",
       "      <td>23</td>\n",
       "      <td>118</td>\n",
       "      <td>0.63</td>\n",
       "      <td>28.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>CAR46496.1</td>\n",
       "      <td>24.503</td>\n",
       "      <td>151</td>\n",
       "      <td>93</td>\n",
       "      <td>4</td>\n",
       "      <td>442</td>\n",
       "      <td>858</td>\n",
       "      <td>91</td>\n",
       "      <td>232</td>\n",
       "      <td>0.90</td>\n",
       "      <td>28.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>CAR42262.1</td>\n",
       "      <td>43.478</td>\n",
       "      <td>23</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>787</td>\n",
       "      <td>855</td>\n",
       "      <td>191</td>\n",
       "      <td>213</td>\n",
       "      <td>1.10</td>\n",
       "      <td>27.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>CAR45487.1</td>\n",
       "      <td>31.915</td>\n",
       "      <td>47</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>295</td>\n",
       "      <td>435</td>\n",
       "      <td>131</td>\n",
       "      <td>177</td>\n",
       "      <td>2.10</td>\n",
       "      <td>26.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>CAR46650.1</td>\n",
       "      <td>33.333</td>\n",
       "      <td>33</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>566</td>\n",
       "      <td>468</td>\n",
       "      <td>529</td>\n",
       "      <td>561</td>\n",
       "      <td>3.90</td>\n",
       "      <td>25.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>wildtype</td>\n",
       "      <td>CAR42804.1</td>\n",
       "      <td>30.233</td>\n",
       "      <td>43</td>\n",
       "      <td>25</td>\n",
       "      <td>1</td>\n",
       "      <td>760</td>\n",
       "      <td>873</td>\n",
       "      <td>21</td>\n",
       "      <td>63</td>\n",
       "      <td>4.20</td>\n",
       "      <td>23.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      query     subject  pc_identity  aln_length  mismatches  gaps_opened  \\\n",
       "0  wildtype  CAR42190.1      100.000         287           0            0   \n",
       "1  wildtype  CAR41317.1       29.412         102          62            3   \n",
       "2  wildtype  CAR46496.1       24.503         151          93            4   \n",
       "3  wildtype  CAR42262.1       43.478          23          13            0   \n",
       "4  wildtype  CAR45487.1       31.915          47          32            0   \n",
       "5  wildtype  CAR46650.1       33.333          33          22            0   \n",
       "6  wildtype  CAR42804.1       30.233          43          25            1   \n",
       "\n",
       "   query_start  query_end  subject_start  subject_end  e_value  bitscore  \n",
       "0           61        921              1          287     0.00     595.0  \n",
       "1           76        369             23          118     0.63      28.1  \n",
       "2          442        858             91          232     0.90      28.1  \n",
       "3          787        855            191          213     1.10      27.7  \n",
       "4          295        435            131          177     2.10      26.6  \n",
       "5          566        468            529          561     3.90      25.8  \n",
       "6          760        873             21           63     4.20      23.9  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read TBLASTX output and add headers\n",
    "results = pd.read_csv(os.path.join(\"output\", \"wildtype_blastx_ref_protein.tab\"), sep=\"\\t\", header=None)\n",
    "results.columns = headers\n",
    "\n",
    "# Show results\n",
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert-success\">\n",
    "The `BLASTX` search showed similarity of the wildtype sequence to seven candidate proteins, but only one candidate, `CAR42190.1`, had more than 30% sequence identity over a considerable part of the sequence.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 01c - Identifying a putative regulatory region\n",
    "\n",
    "To identify a putative regulatory region, the reference genome is inspected in the region of the `TBLASTX` hit and annotated location of `CAR42190.1`, and the region immediately upstream of the operon containing this location extracted as a nucleotide sequence in FASTA format.\n",
    "\n",
    "* **Ask the learners to start `Artemis` (at the command-line)**\n",
    "* **Ask the learners to load the `data/GCA_000069965.1_ASM6996v1_genomic.gbk` file and navigate to the region of the `TBLASTX` match.**\n",
    "\n",
    "![Artemis region match](images/XX-01_region.png)\n",
    "\n",
    "* **Ask the learners to get information about the annotated feature immediately downstream (this is `PMI0999`, and is their `BLASTX` hit). Note that this is confirmation of the `BLASTX` result.**\n",
    "\n",
    "![Artemis feature](images/XX-02_feature.png)\n",
    "\n",
    "* **Ask the learners to select the genomic region on the forward strand between the two features `PMI0998` and `PMI0999`: `1062938..1063081`**\n",
    "\n",
    "![Artemis selection](images/XX-03_select.png)\n",
    "\n",
    "* **Ask the learners to write that sequence out to `output/putative_regulatory_region.fasta`.**\n",
    "\n",
    "![Artemis menu](images/XX-04_menu.png)\n",
    "![Artemis write](images/XX-05_write.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Ask learners to write up their work.**\n",
    "\n",
    "<p></p>\n",
    "<div class=\"alert-success\">\n",
    "The region of the `TBLASTX` match was inspected using `Artemis` and found to include the region annotated with `PMI0999`/`CAR42190`, the hit identified by `BLASTX`. The intergenic region upstream of `PMI0999` was extracted as a FASTA sequence file in `output/putative_regulatory_region.fasta`.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 02 - Identifying homologues in public databases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 02a - BLAST/NCBI\n",
    "\n",
    "The first thing to do is query nr with our wildtype protein sequence, using `BLASTP`. This will be done remotely, and can be done from the notebook or through the web browser. The web browser seems to run *much* faster, in testing.\n",
    "\n",
    "* **Ask the learners to submit the `wildtype_aa.fasta` sequence as a query against the remote `nr` database, restricting their query to `Proteus` (this will save some time)**\n",
    "* **Remind learners that the `wildtype_aa` sequence is still in memory, so they can use that**\n",
    "* **Even restricting to `Proteus`, the remote search will take a couple of minutes - use this to write up the notes**\n",
    "* **The web browser option will give an alignment file almost immediately, and is probably the safe route to take when presenting the material**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert-success\">\n",
    "The translated wildtype sequence was used as a remote `BLASTP` query against the `nr` database at NCBI. The query was restricted to organisms in the Genus *Proteus*.\n",
    "</div>\n",
    "\n",
    "* **Option 1: In the notebook**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Submit wildtype_aa query sequence to NCBI nr BLASTP, restricting\n",
    "# to enterobacteriales\n",
    "result_handle = NCBIWWW.qblast(\"blastp\", \"nr\", wildtype_aa.format(\"fasta\"),\n",
    "                               entrez_query=\"proteus[ORGN]\",\n",
    "                               format_type=\"Text\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<p><!--\n",
      "QBlastInfoBegin\n",
      "\tStatus=READY\n",
      "QBlastInfoEnd\n",
      "--><p>\n",
      "<PRE>\n",
      "BLASTP 2.6.1+\n",
      "Reference: Stephen F. Altschul, Thomas L. Madden, Alejandro\n",
      "A. Schaffer, Jinghui Zhang, Zheng Zhang, Webb Miller, and\n",
      "David J. Lipman (1997), \"Gapped BLAST and PSI-BLAST: a new\n",
      "generation of protein database search programs\", Nucleic\n",
      "Acids Res. 25:3389-3402.\n",
      "\n",
      "\n",
      "Reference for compositional score matrix adjustment: Stephen\n",
      "F. Altschul, John C. Wootton, E. Michael Gertz, Richa\n",
      "Agarwala, Aleksandr Morgulis, Alejandro A. Schaffer, and\n",
      "Yi-Kuo Yu (2005) \"Protein database searches using\n",
      "compositionally adjusted substitution matrices\", FEBS J.\n",
      "272:5101-5109.\n",
      "\n",
      "\n",
      "RID: CDFC2WWF016\n",
      "\n",
      "\n",
      "Database: All non-redundant GenBank CDS\n",
      "translations+PDB+SwissProt+PIR+PRF excluding environmental samples\n",
      "from WGS projects\n",
      "           115,570,790 sequences; 42,364,384,627 total letters\n",
      "Query= translated_wildtype wildtype lipase protein from Proteus mirabilis\n",
      "\n",
      "Length=307\n",
      "\n",
      "\n",
      "                                                                   Score     E\n",
      "Sequences producing significant alignments:                       (Bits)  Value\n",
      "\n",
      "4GW3_A  Chain A, Crystal Structure Of The Lipase From Proteus ...  639     0.0   \n",
      "3W9U_A  Chain A, Crystal Structure Of Lipk107                      603     0.0   \n",
      "WP_004247922.1  MULTISPECIES: lipase [Proteus]                     597     0.0   \n",
      "WP_063215726.1  alpha/beta hydrolase [Proteus mirabilis]           595     0.0   \n",
      "WP_046334983.1  alpha/beta hydrolase [Proteus mirabilis]           595     0.0   \n",
      "WP_049203868.1  alpha/beta hydrolase [Proteus mirabilis]           594     0.0   \n",
      "WP_004242549.1  MULTISPECIES: lipase [Proteus]                     594     0.0   \n",
      "4HS9_A  Chain A, Methanol Tolerant Mutant Of The Proteus Mirab...  585     0.0   \n",
      "WP_060559447.1  alpha/beta hydrolase [Proteus mirabilis]           581     0.0   \n",
      "OOQ53411.1  alpha/beta hydrolase [Proteus mirabilis]               580     0.0   \n",
      "WP_072070422.1  alpha/beta hydrolase [Proteus vulgaris]            499     2e-180\n",
      "WP_023582422.1  lipase [Proteus hauseri]                           498     1e-179\n",
      "WP_075672614.1  alpha/beta hydrolase [Proteus sp. H24]             494     2e-178\n",
      "WP_036937497.1  alpha/beta hydrolase [Proteus vulgaris]            494     3e-178\n",
      "AAB01071.1  alkaline lipase [Proteus vulgaris]                     493     6e-178\n",
      "AFB78082.1  lipase [Proteus sp. enrichment culture clone K19]      488     4e-176\n",
      "ACM67042.1  lipase [Proteus vulgaris]                              486     3e-175\n",
      "ADV92635.1  alkaline lipase [Proteus sp. SW1]                      484     3e-174\n",
      "WP_072062608.1  alpha/beta hydrolase [Proteus vulgaris]            480     7e-173\n",
      "WP_064719666.1  alpha/beta hydrolase [Proteus hauseri]             470     1e-168\n",
      "EEG83604.1  triacylglycerol lipase family protein [Proteus pen...  286     8e-98 \n",
      "EEG83605.1  hypothetical protein PROPEN_04374 [Proteus penneri...  128     2e-37 \n",
      "WP_014837821.1  MULTISPECIES: acyl-CoA esterase [Enterobactera...  36.6    0.021 \n",
      "WP_036938484.1  acyl-CoA esterase [Proteus vulgaris]               35.8    0.040 \n",
      "WP_072070221.1  alpha/beta hydrolase [Proteus vulgaris]            34.3    0.12  \n",
      "WP_064719330.1  acyl-CoA esterase [Proteus hauseri]                33.5    0.24  \n",
      "WP_075671659.1  alpha/beta hydrolase [Proteus sp. H24]             33.5    0.25  \n",
      "WP_072062930.1  alpha/beta hydrolase [Proteus vulgaris]            33.1    0.37  \n",
      "EEG84125.1  PGAP1-like protein [Proteus penneri ATCC 35198]        32.3    0.50  \n",
      "WP_023581172.1  hydrolase [Proteus hauseri]                        32.3    0.60  \n",
      "AMR60768.1  replicase [Proteus mirabilis]                          30.4    2.5   \n",
      "WP_072196818.1  hypothetical protein [Proteus mirabilis]           26.6    9.3   \n",
      "\n",
      "ALIGNMENTS\n",
      ">4GW3_A Chain A, Crystal Structure Of The Lipase From Proteus Mirabilis\n",
      " 4GXN_A Chain A, Diethylphosphonate Inhibited Structure Of The Proteus \n",
      "Mirabilis Lipase\n",
      "Length=307\n",
      "\n",
      " Score = 639 bits (1647),  Expect = 0.0, Method: Compositional matrix adjust.\n",
      " Identities = 307/307 (100%), Positives = 307/307 (100%), Gaps = 0/307 (0%)\n",
      "\n",
      "Query  1    MGSSHHHHHHSSGLVPRGSHMSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVF  60\n",
      "            MGSSHHHHHHSSGLVPRGSHMSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVF\n",
      "Sbjct  1    MGSSHHHHHHSSGLVPRGSHMSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVF  60\n",
      "\n",
      "Query  61   TASLSAFNSNEVRGKQLWQFVQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVT  120\n",
      "            TASLSAFNSNEVRGKQLWQFVQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVT\n",
      "Sbjct  61   TASLSAFNSNEVRGKQLWQFVQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVT  120\n",
      "\n",
      "Query  121  SINGVNHGSEIADLYRRIMRKDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESL  180\n",
      "            SINGVNHGSEIADLYRRIMRKDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESL\n",
      "Sbjct  121  SINGVNHGSEIADLYRRIMRKDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESL  180\n",
      "\n",
      "Query  181  TTEQVTEFNNKYPQALPKTPGGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMR  240\n",
      "            TTEQVTEFNNKYPQALPKTPGGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMR\n",
      "Sbjct  181  TTEQVTEFNNKYPQALPKTPGGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMR  240\n",
      "\n",
      "Query  241  VLNTFFTEKQNDGLVGRSSMRLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAK  300\n",
      "            VLNTFFTEKQNDGLVGRSSMRLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAK\n",
      "Sbjct  241  VLNTFFTEKQNDGLVGRSSMRLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAK  300\n",
      "\n",
      "Query  301  YLASKQL  307\n",
      "            YLASKQL\n",
      "Sbjct  301  YLASKQL  307\n",
      "\n",
      "\n",
      ">3W9U_A Chain A, Crystal Structure Of Lipk107\n",
      "Length=290\n",
      "\n",
      " Score = 603 bits (1554),  Expect = 0.0, Method: Compositional matrix adjust.\n",
      " Identities = 290/290 (100%), Positives = 290/290 (100%), Gaps = 0/290 (0%)\n",
      "\n",
      "Query  18   GSHMSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQL  77\n",
      "            GSHMSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQL\n",
      "Sbjct  1    GSHMSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQL  60\n",
      "\n",
      "Query  78   WQFVQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRR  137\n",
      "            WQFVQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRR\n",
      "Sbjct  61   WQFVQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRR  120\n",
      "\n",
      "Query  138  IMRKDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALP  197\n",
      "            IMRKDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALP\n",
      "Sbjct  121  IMRKDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALP  180\n",
      "\n",
      "Query  198  KTPGGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGR  257\n",
      "            KTPGGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGR\n",
      "Sbjct  181  KTPGGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGR  240\n",
      "\n",
      "Query  258  SSMRLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            SSMRLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL\n",
      "Sbjct  241  SSMRLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  290\n",
      "\n",
      "\n",
      ">WP_004247922.1 MULTISPECIES: lipase [Proteus]\n",
      " ACC76759.1 lipase [Proteus sp. K107]\n",
      " CAR42190.1 putative lipase [Proteus mirabilis HI4320]\n",
      " EKA97054.1 hypothetical protein HMPREF1310_02199 [Proteus mirabilis WGLW4]\n",
      " EKB01462.1 hypothetical protein HMPREF1311_00993 [Proteus mirabilis WGLW6]\n",
      " AGS59531.1 lipase [Proteus mirabilis BB2000]\n",
      " KGA89882.1 alpha/beta hydrolase fold family protein [Proteus mirabilis]\n",
      " KGQ17691.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " KGY38074.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " KGY42161.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " KGZ39216.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " ALE22010.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " ALE25151.1 alpha/beta hydrolase [Proteus vulgaris]\n",
      " KSA05190.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " KSW21244.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " KXC01436.1 putative lactonizing lipase [Proteus mirabilis]\n",
      " AND13297.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " OAH93856.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " OAS29122.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " OAS38840.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " SDC88067.1 triacylglycerol lipase [Proteus mirabilis]\n",
      " SFH21238.1 triacylglycerol lipase [Proteus mirabilis]\n",
      " OIK52547.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      "Length=287\n",
      "\n",
      " Score = 597 bits (1538),  Expect = 0.0, Method: Compositional matrix adjust.\n",
      " Identities = 287/287 (100%), Positives = 287/287 (100%), Gaps = 0/287 (0%)\n",
      "\n",
      "Query  21   MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  80\n",
      "            MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF\n",
      "Sbjct  1    MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  60\n",
      "\n",
      "Query  81   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  140\n",
      "            VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR\n",
      "Sbjct  61   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  120\n",
      "\n",
      "Query  141  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  200\n",
      "            KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP\n",
      "Sbjct  121  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  180\n",
      "\n",
      "Query  201  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  260\n",
      "            GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM\n",
      "Sbjct  181  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  240\n",
      "\n",
      "Query  261  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL\n",
      "Sbjct  241  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  287\n",
      "\n",
      "\n",
      ">WP_063215726.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " KZE57832.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      "Length=287\n",
      "\n",
      " Score = 595 bits (1535),  Expect = 0.0, Method: Compositional matrix adjust.\n",
      " Identities = 286/287 (99%), Positives = 287/287 (100%), Gaps = 0/287 (0%)\n",
      "\n",
      "Query  21   MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  80\n",
      "            MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF\n",
      "Sbjct  1    MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  60\n",
      "\n",
      "Query  81   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  140\n",
      "            VQTLLQETQAKK+NFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR\n",
      "Sbjct  61   VQTLLQETQAKKLNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  120\n",
      "\n",
      "Query  141  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  200\n",
      "            KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP\n",
      "Sbjct  121  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  180\n",
      "\n",
      "Query  201  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  260\n",
      "            GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM\n",
      "Sbjct  181  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  240\n",
      "\n",
      "Query  261  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL\n",
      "Sbjct  241  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  287\n",
      "\n",
      "\n",
      ">WP_046334983.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " KKC58863.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      "Length=287\n",
      "\n",
      " Score = 595 bits (1534),  Expect = 0.0, Method: Compositional matrix adjust.\n",
      " Identities = 286/287 (99%), Positives = 287/287 (100%), Gaps = 0/287 (0%)\n",
      "\n",
      "Query  21   MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  80\n",
      "            MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSA+NSNEVRGKQLWQF\n",
      "Sbjct  1    MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAYNSNEVRGKQLWQF  60\n",
      "\n",
      "Query  81   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  140\n",
      "            VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR\n",
      "Sbjct  61   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  120\n",
      "\n",
      "Query  141  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  200\n",
      "            KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP\n",
      "Sbjct  121  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  180\n",
      "\n",
      "Query  201  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  260\n",
      "            GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM\n",
      "Sbjct  181  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  240\n",
      "\n",
      "Query  261  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL\n",
      "Sbjct  241  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  287\n",
      "\n",
      "\n",
      ">WP_049203868.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      "Length=287\n",
      "\n",
      " Score = 594 bits (1531),  Expect = 0.0, Method: Compositional matrix adjust.\n",
      " Identities = 286/287 (99%), Positives = 286/287 (99%), Gaps = 0/287 (0%)\n",
      "\n",
      "Query  21   MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  80\n",
      "            MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF\n",
      "Sbjct  1    MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  60\n",
      "\n",
      "Query  81   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  140\n",
      "            VQ LLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR\n",
      "Sbjct  61   VQILLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  120\n",
      "\n",
      "Query  141  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  200\n",
      "            KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP\n",
      "Sbjct  121  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  180\n",
      "\n",
      "Query  201  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  260\n",
      "            GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM\n",
      "Sbjct  181  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  240\n",
      "\n",
      "Query  261  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL\n",
      "Sbjct  241  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  287\n",
      "\n",
      "\n",
      ">WP_004242549.1 MULTISPECIES: lipase [Proteus]\n",
      " EEI49517.1 hydrolase, alpha/beta domain protein [Proteus mirabilis ATCC \n",
      "29906]\n",
      " KGY46371.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " KGZ02169.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " KSW19467.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " OFV22723.1 alpha/beta hydrolase [Proteus sp. HMSC14B05]\n",
      " OHY46277.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      "Length=287\n",
      "\n",
      " Score = 594 bits (1531),  Expect = 0.0, Method: Compositional matrix adjust.\n",
      " Identities = 286/287 (99%), Positives = 286/287 (99%), Gaps = 0/287 (0%)\n",
      "\n",
      "Query  21   MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  80\n",
      "            MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF\n",
      "Sbjct  1    MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  60\n",
      "\n",
      "Query  81   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  140\n",
      "            VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR\n",
      "Sbjct  61   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  120\n",
      "\n",
      "Query  141  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  200\n",
      "            KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPK P\n",
      "Sbjct  121  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKIP  180\n",
      "\n",
      "Query  201  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  260\n",
      "            GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM\n",
      "Sbjct  181  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  240\n",
      "\n",
      "Query  261  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL\n",
      "Sbjct  241  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  287\n",
      "\n",
      "\n",
      ">4HS9_A Chain A, Methanol Tolerant Mutant Of The Proteus Mirabilis Lipase\n",
      "Length=307\n",
      "\n",
      " Score = 585 bits (1507),  Expect = 0.0, Method: Compositional matrix adjust.\n",
      " Identities = 293/307 (95%), Positives = 297/307 (97%), Gaps = 0/307 (0%)\n",
      "\n",
      "Query  1    MGSSHHHHHHSSGLVPRGSHMSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVF  60\n",
      "            MGSSHHHHHHSSGLVPRGSHMSTKYPIVLVHGLAGF+EIVGFPYFYGIADAL QDGHQVF\n",
      "Sbjct  1    MGSSHHHHHHSSGLVPRGSHMSTKYPIVLVHGLAGFSEIVGFPYFYGIADALTQDGHQVF  60\n",
      "\n",
      "Query  61   TASLSAFNSNEVRGKQLWQFVQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVT  120\n",
      "            TASLSAFNSNEVRGKQLWQFVQT+LQETQ KKVNFIGHSQGPLACRYVAANYPDSVASVT\n",
      "Sbjct  61   TASLSAFNSNEVRGKQLWQFVQTILQETQTKKVNFIGHSQGPLACRYVAANYPDSVASVT  120\n",
      "\n",
      "Query  121  SINGVNHGSEIADLYRRIMRKDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESL  180\n",
      "            SINGVNHGSEIADLYRRI+RKDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESL\n",
      "Sbjct  121  SINGVNHGSEIADLYRRIIRKDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESL  180\n",
      "\n",
      "Query  181  TTEQVTEFNNKYPQALPKTPGGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMR  240\n",
      "            TTEQVTEFNNKYPQALPKTP GEGDEIVNGVHYYCFGSYIQ LIAGE GNLLDPTHAAMR\n",
      "Sbjct  181  TTEQVTEFNNKYPQALPKTPCGEGDEIVNGVHYYCFGSYIQELIAGENGNLLDPTHAAMR  240\n",
      "\n",
      "Query  241  VLNTFFTEKQNDGLVGRSSMRLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAK  300\n",
      "            VLNT FTEKQNDGLVGR SMRLGKLIKDDYAQDH DMVNQVAGLV YNE+IVAIYT HAK\n",
      "Sbjct  241  VLNTLFTEKQNDGLVGRCSMRLGKLIKDDYAQDHFDMVNQVAGLVSYNENIVAIYTLHAK  300\n",
      "\n",
      "Query  301  YLASKQL  307\n",
      "            YLASKQL\n",
      "Sbjct  301  YLASKQL  307\n",
      "\n",
      "\n",
      ">WP_060559447.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      " KSX93692.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      "Length=287\n",
      "\n",
      " Score = 581 bits (1497),  Expect = 0.0, Method: Compositional matrix adjust.\n",
      " Identities = 279/287 (97%), Positives = 282/287 (98%), Gaps = 0/287 (0%)\n",
      "\n",
      "Query  21   MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  80\n",
      "            MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF\n",
      "Sbjct  1    MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  60\n",
      "\n",
      "Query  81   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  140\n",
      "            VQTLL+ETQAKKVNFIGHSQGPLACRYVAA YPDSVASVTSINGVNHGSEIADL+RRIMR\n",
      "Sbjct  61   VQTLLKETQAKKVNFIGHSQGPLACRYVAAKYPDSVASVTSINGVNHGSEIADLFRRIMR  120\n",
      "\n",
      "Query  141  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  200\n",
      "            KDS+PEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPK P\n",
      "Sbjct  121  KDSVPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKIP  180\n",
      "\n",
      "Query  201  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  260\n",
      "            GGEGDEIVNGVHYY FGSYIQGLIAGEKGNLLDPTHAAMRVLN FF EKQNDGLVGRSSM\n",
      "Sbjct  181  GGEGDEIVNGVHYYSFGSYIQGLIAGEKGNLLDPTHAAMRVLNAFFVEKQNDGLVGRSSM  240\n",
      "\n",
      "Query  261  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL\n",
      "Sbjct  241  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  287\n",
      "\n",
      "\n",
      ">OOQ53411.1 alpha/beta hydrolase [Proteus mirabilis]\n",
      "Length=287\n",
      "\n",
      " Score = 580 bits (1496),  Expect = 0.0, Method: Compositional matrix adjust.\n",
      " Identities = 279/287 (97%), Positives = 282/287 (98%), Gaps = 0/287 (0%)\n",
      "\n",
      "Query  21   MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  80\n",
      "            MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF\n",
      "Sbjct  1    MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  60\n",
      "\n",
      "Query  81   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  140\n",
      "            VQTLL+ETQAKKVNFIGHSQGPLACRYVAA YPDSVASVTSINGVNHGSEIADL+RRIMR\n",
      "Sbjct  61   VQTLLKETQAKKVNFIGHSQGPLACRYVAAKYPDSVASVTSINGVNHGSEIADLFRRIMR  120\n",
      "\n",
      "Query  141  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  200\n",
      "            KDS+PEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPK P\n",
      "Sbjct  121  KDSVPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKIP  180\n",
      "\n",
      "Query  201  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  260\n",
      "            GGEGDEIVNGVHYY FGSYIQGLIAGEKGNLLDPTHAAMRVLN FF EKQNDGLVGRSSM\n",
      "Sbjct  181  GGEGDEIVNGVHYYSFGSYIQGLIAGEKGNLLDPTHAAMRVLNAFFIEKQNDGLVGRSSM  240\n",
      "\n",
      "Query  261  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL\n",
      "Sbjct  241  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  287\n",
      "\n",
      "\n",
      ">WP_072070422.1 alpha/beta hydrolase [Proteus vulgaris]\n",
      " CRL64835.1 Lactonizing lipase precursor [Proteus vulgaris]\n",
      "Length=287\n",
      "\n",
      " Score = 499 bits (1285),  Expect = 2e-180, Method: Compositional matrix adjust.\n",
      " Identities = 232/287 (81%), Positives = 263/287 (92%), Gaps = 0/287 (0%)\n",
      "\n",
      "Query  21   MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  80\n",
      "            MST YPI+LVHGL+GF++IVG+PYFYGIADAL ++GHQVFTASLSAFNSNEVRG+QLW+F\n",
      "Sbjct  1    MSTTYPIILVHGLSGFDDIVGYPYFYGIADALEKEGHQVFTASLSAFNSNEVRGEQLWEF  60\n",
      "\n",
      "Query  81   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  140\n",
      "            VQ +L+ET+AKKVN IGHSQGPLACRYVAA +  ++ASVTS+NGVNHGSEIADL RRIMR\n",
      "Sbjct  61   VQKVLKETKAKKVNLIGHSQGPLACRYVAAKHAKNIASVTSVNGVNHGSEIADLVRRIMR  120\n",
      "\n",
      "Query  141  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  200\n",
      "            KDS+PEYI + V+ A GTIISTFSG+RG+PQDAIAALE+LTTE VTEFN KYPQ LP+  \n",
      "Sbjct  121  KDSVPEYIADAVMKAIGTIISTFSGNRGNPQDAIAALEALTTENVTEFNKKYPQGLPEVR  180\n",
      "\n",
      "Query  201  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  260\n",
      "            GGEG E+VNGVHYY FGSYIQGLIAGEKGNLLDPTHAAMRVL+ FF E+QNDGLVGR+SM\n",
      "Sbjct  181  GGEGKEVVNGVHYYSFGSYIQGLIAGEKGNLLDPTHAAMRVLSAFFIERQNDGLVGRTSM  240\n",
      "\n",
      "Query  261  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            RLGKLIKDDYA+DHIDMVNQVAGLVG  EDI+AIYT HAK+LASK+L\n",
      "Sbjct  241  RLGKLIKDDYAEDHIDMVNQVAGLVGRGEDIIAIYTNHAKFLASKKL  287\n",
      "\n",
      "\n",
      ">WP_023582422.1 lipase [Proteus hauseri]\n",
      " EST59074.1 lipase [Proteus hauseri ZMd44]\n",
      "Length=287\n",
      "\n",
      " Score = 498 bits (1281),  Expect = 1e-179, Method: Compositional matrix adjust.\n",
      " Identities = 230/287 (80%), Positives = 262/287 (91%), Gaps = 0/287 (0%)\n",
      "\n",
      "Query  21   MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  80\n",
      "            MST YPIVLVHGL+GF++IVG+PYFYGIADAL +DGH+VFTASLSAFNSNEVRG+QLW+F\n",
      "Sbjct  1    MSTTYPIVLVHGLSGFDDIVGYPYFYGIADALEKDGHKVFTASLSAFNSNEVRGEQLWEF  60\n",
      "\n",
      "Query  81   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  140\n",
      "            VQ +L+ET+AKKVN IGHSQGPLACRYVAA +  S+ASVTS+NGVNHGSEIADL RRIMR\n",
      "Sbjct  61   VQKVLKETKAKKVNLIGHSQGPLACRYVAAKHAKSIASVTSVNGVNHGSEIADLVRRIMR  120\n",
      "\n",
      "Query  141  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  200\n",
      "            KDS+PEYI + V+ A GTIISTFSG+RG+PQDA+AALE+LTTE VTEFN KYPQ LP   \n",
      "Sbjct  121  KDSVPEYIADAVMKAIGTIISTFSGNRGNPQDAVAALEALTTENVTEFNKKYPQGLPAVR  180\n",
      "\n",
      "Query  201  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  260\n",
      "            GGEG EIVNGVHYY FGSYIQGLIAGEKGNLLDPTHAAMR+L+ FF+E++NDGLVGR+SM\n",
      "Sbjct  181  GGEGKEIVNGVHYYSFGSYIQGLIAGEKGNLLDPTHAAMRILSAFFSERENDGLVGRTSM  240\n",
      "\n",
      "Query  261  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            RLGKLIKDDYA+DH+DMVNQVAGLVG  EDI+AIYT HA +LASK+L\n",
      "Sbjct  241  RLGKLIKDDYAEDHLDMVNQVAGLVGRGEDIIAIYTNHANFLASKKL  287\n",
      "\n",
      "\n",
      ">WP_075672614.1 alpha/beta hydrolase [Proteus sp. H24]\n",
      "Length=287\n",
      "\n",
      " Score = 494 bits (1272),  Expect = 2e-178, Method: Compositional matrix adjust.\n",
      " Identities = 229/287 (80%), Positives = 260/287 (91%), Gaps = 0/287 (0%)\n",
      "\n",
      "Query  21   MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  80\n",
      "            MST YPIVLVHGL+GF++IVG+PYFYGIADAL +DGH+VFTASLSAFNSNEVRG+QLW+F\n",
      "Sbjct  1    MSTTYPIVLVHGLSGFDDIVGYPYFYGIADALEKDGHKVFTASLSAFNSNEVRGEQLWEF  60\n",
      "\n",
      "Query  81   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  140\n",
      "            VQ +L+ET+ KKVN IGHSQGPLACRYVAA +  S+ASVTS+NGVNHGSEIADL RRIMR\n",
      "Sbjct  61   VQKILKETKVKKVNLIGHSQGPLACRYVAAKHAKSIASVTSVNGVNHGSEIADLVRRIMR  120\n",
      "\n",
      "Query  141  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  200\n",
      "            KDS+PEYI + V+ A GTIISTFSG+RG+PQDAIAALE+LTTE V EFN KYPQ LP   \n",
      "Sbjct  121  KDSVPEYIADAVMKAIGTIISTFSGNRGNPQDAIAALEALTTENVMEFNKKYPQGLPAIR  180\n",
      "\n",
      "Query  201  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  260\n",
      "            GGEG E+VNGVHYY FGSYIQGLIAGEKGNLLDPTHAAMRVL+ FF+E++NDGLVGR+SM\n",
      "Sbjct  181  GGEGKEVVNGVHYYSFGSYIQGLIAGEKGNLLDPTHAAMRVLSAFFSERENDGLVGRTSM  240\n",
      "\n",
      "Query  261  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            RLGKLIKDDYA+DH+DMVNQVAGLVG  EDI+AIYT HA +LASK+L\n",
      "Sbjct  241  RLGKLIKDDYAEDHLDMVNQVAGLVGRGEDIIAIYTNHANFLASKKL  287\n",
      "\n",
      "\n",
      ">WP_036937497.1 alpha/beta hydrolase [Proteus vulgaris]\n",
      " KGA60007.1 alpha/beta hydrolase fold family protein [Proteus vulgaris]\n",
      "Length=287\n",
      "\n",
      " Score = 494 bits (1271),  Expect = 3e-178, Method: Compositional matrix adjust.\n",
      " Identities = 227/287 (79%), Positives = 262/287 (91%), Gaps = 0/287 (0%)\n",
      "\n",
      "Query  21   MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  80\n",
      "            MST YPI+LVHGL+GF++IVG+PYFYGIADAL ++G++VFTASLSAFNSNEVRG+QLW+F\n",
      "Sbjct  1    MSTTYPIILVHGLSGFDDIVGYPYFYGIADALEKEGNKVFTASLSAFNSNEVRGEQLWEF  60\n",
      "\n",
      "Query  81   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  140\n",
      "            VQ +L+ET+ +KVN IGHSQGPLACRYVAA Y  ++ASVTS+NGVNHGSEIADL RRIMR\n",
      "Sbjct  61   VQKILKETKVRKVNLIGHSQGPLACRYVAAKYAKNIASVTSVNGVNHGSEIADLVRRIMR  120\n",
      "\n",
      "Query  141  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  200\n",
      "            KDS+PEYIV+ V+ A GTIISTFSG+RG+PQDAIAALE+LTTE VTEFN KYPQ LP+  \n",
      "Sbjct  121  KDSVPEYIVDAVMKAIGTIISTFSGNRGNPQDAIAALEALTTENVTEFNKKYPQGLPELR  180\n",
      "\n",
      "Query  201  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  260\n",
      "            GGEG E+VNGVHYY FGSYIQGLIAGEK NLLDPTHAAMRVL+ FF E+QNDGLVGR+SM\n",
      "Sbjct  181  GGEGKEVVNGVHYYSFGSYIQGLIAGEKSNLLDPTHAAMRVLSAFFIERQNDGLVGRTSM  240\n",
      "\n",
      "Query  261  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            RLGKL+KDDYA+DH+DMVNQVAGLVG  EDI+AIYT HAK+LASK+L\n",
      "Sbjct  241  RLGKLVKDDYAEDHLDMVNQVAGLVGRGEDIIAIYTNHAKFLASKKL  287\n",
      "\n",
      "\n",
      ">AAB01071.1 alkaline lipase [Proteus vulgaris]\n",
      "Length=290\n",
      "\n",
      " Score = 493 bits (1270),  Expect = 6e-178, Method: Compositional matrix adjust.\n",
      " Identities = 230/288 (80%), Positives = 261/288 (91%), Gaps = 0/288 (0%)\n",
      "\n",
      "Query  20   HMSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQ  79\n",
      "            +MST YPIVLVHGL+GF++IVG+PYFYGI DAL +DGH+VFTASLSAFNSNEVRG+QLW+\n",
      "Sbjct  3    NMSTTYPIVLVHGLSGFDDIVGYPYFYGIRDALEKDGHKVFTASLSAFNSNEVRGEQLWE  62\n",
      "\n",
      "Query  80   FVQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIM  139\n",
      "            FVQ +L+ET+AKKVN IGHSQGPLACRYVAA +  ++ASVTS+NGVNHGSEIADL RRIM\n",
      "Sbjct  63   FVQKVLKETKAKKVNLIGHSQGPLACRYVAAKHAKNIASVTSVNGVNHGSEIADLVRRIM  122\n",
      "\n",
      "Query  140  RKDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKT  199\n",
      "            RKDS+PEYI + V+ A GTIISTFSG+RG+PQDAIAALE+LTTE V EFN KYPQ LP  \n",
      "Sbjct  123  RKDSVPEYIADAVMKAIGTIISTFSGNRGNPQDAIAALEALTTENVMEFNKKYPQGLPAI  182\n",
      "\n",
      "Query  200  PGGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSS  259\n",
      "             GGEG E+VNGVHYY FGSYIQGLIAGEKGNLLDPTHAAMRVL+ FFTE++NDGLVGR+S\n",
      "Sbjct  183  RGGEGKEVVNGVHYYSFGSYIQGLIAGEKGNLLDPTHAAMRVLSAFFTERENDGLVGRTS  242\n",
      "\n",
      "Query  260  MRLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            MRLGKLIKDDYA+DH+DMVNQVAGLVG  EDIVAIYT HA +LASK+L\n",
      "Sbjct  243  MRLGKLIKDDYAEDHLDMVNQVAGLVGPGEDIVAIYTNHANFLASKKL  290\n",
      "\n",
      "\n",
      ">AFB78082.1 lipase [Proteus sp. enrichment culture clone K19]\n",
      "Length=287\n",
      "\n",
      " Score = 488 bits (1257),  Expect = 4e-176, Method: Compositional matrix adjust.\n",
      " Identities = 226/287 (79%), Positives = 258/287 (90%), Gaps = 0/287 (0%)\n",
      "\n",
      "Query  21   MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  80\n",
      "            MST YPIVLVHGL+GF++IVG+PYFYGIADAL +DGH+VFTASLSAFNSNEVRG+QLW+F\n",
      "Sbjct  1    MSTTYPIVLVHGLSGFDDIVGYPYFYGIADALEKDGHKVFTASLSAFNSNEVRGEQLWEF  60\n",
      "\n",
      "Query  81   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  140\n",
      "            VQ +L+E +AKKVN IGHSQGPLACRYVAA +  S+ASV S+NGVNHGSEIADL RRIMR\n",
      "Sbjct  61   VQKVLKEIKAKKVNLIGHSQGPLACRYVAAKHAKSIASVASVNGVNHGSEIADLVRRIMR  120\n",
      "\n",
      "Query  141  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  200\n",
      "            KDS+PEY  + V+ A GTIISTFSG+ G+PQDA+AALE+LTTE VTEFN KYPQ LP   \n",
      "Sbjct  121  KDSVPEYTADAVMKAIGTIISTFSGNHGNPQDAVAALEALTTENVTEFNKKYPQGLPAVR  180\n",
      "\n",
      "Query  201  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  260\n",
      "            GGEG EIVNGVHYY FGSYIQGLIAGEKGNLLDPTHAAMR+L+ FF+E++NDGLVGR+SM\n",
      "Sbjct  181  GGEGKEIVNGVHYYSFGSYIQGLIAGEKGNLLDPTHAAMRILSAFFSERENDGLVGRTSM  240\n",
      "\n",
      "Query  261  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            RLGKLIKDDYA+DH+DMVNQVAGLVG  EDI+AIYT HA +LASK+L\n",
      "Sbjct  241  RLGKLIKDDYAEDHLDMVNQVAGLVGRGEDIIAIYTNHANFLASKKL  287\n",
      "\n",
      "\n",
      ">ACM67042.1 lipase [Proteus vulgaris]\n",
      "Length=287\n",
      "\n",
      " Score = 486 bits (1252),  Expect = 3e-175, Method: Compositional matrix adjust.\n",
      " Identities = 227/287 (79%), Positives = 257/287 (90%), Gaps = 0/287 (0%)\n",
      "\n",
      "Query  21   MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  80\n",
      "            MST YPIVLVHGL+GF+ +VG+PYFYGIADAL +DGH+VFTASLSAFN+NEVRG+QLW F\n",
      "Sbjct  1    MSTTYPIVLVHGLSGFDNVVGYPYFYGIADALEKDGHKVFTASLSAFNANEVRGEQLWGF  60\n",
      "\n",
      "Query  81   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  140\n",
      "            VQ +L+ET+AKKVN IGHSQGPLACRYVAA +  ++ASVTSINGVNHGSEIADL RRI+R\n",
      "Sbjct  61   VQKVLKETKAKKVNLIGHSQGPLACRYVAAKHAKNIASVTSINGVNHGSEIADLVRRIVR  120\n",
      "\n",
      "Query  141  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  200\n",
      "            KDS+PEYI + V+ A GTIISTFSGHR +PQDA+AALE+LTTE VTEFN KYPQ LP   \n",
      "Sbjct  121  KDSVPEYIADAVMKAIGTIISTFSGHRSNPQDAVAALEALTTENVTEFNKKYPQGLPAIR  180\n",
      "\n",
      "Query  201  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  260\n",
      "            GGEG E+VNGV+YY FGSYIQGLI GEKGNLLDPTHAAMRVL+ FFTE +NDGLVGR+SM\n",
      "Sbjct  181  GGEGKEVVNGVYYYSFGSYIQGLIVGEKGNLLDPTHAAMRVLSAFFTEHENDGLVGRTSM  240\n",
      "\n",
      "Query  261  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            RLGKLIKDDYA+DH+DMVNQVAGLVG  EDIVAIYT HA +LASK+L\n",
      "Sbjct  241  RLGKLIKDDYAEDHLDMVNQVAGLVGRGEDIVAIYTNHANFLASKKL  287\n",
      "\n",
      "\n",
      ">ADV92635.1 alkaline lipase [Proteus sp. SW1]\n",
      "Length=287\n",
      "\n",
      " Score = 484 bits (1245),  Expect = 3e-174, Method: Compositional matrix adjust.\n",
      " Identities = 226/287 (79%), Positives = 256/287 (89%), Gaps = 0/287 (0%)\n",
      "\n",
      "Query  21   MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  80\n",
      "            M T YPIVLVHGL+GF+ +VG+PYFYGIADAL +DGH+VFTASLSAFN+NEVRG+QLW F\n",
      "Sbjct  1    MPTTYPIVLVHGLSGFDNVVGYPYFYGIADALEKDGHKVFTASLSAFNANEVRGEQLWGF  60\n",
      "\n",
      "Query  81   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  140\n",
      "            VQ +L+ET+AKKVN IGHSQGPLACRYVAA +  ++ASVTSINGVNHGSEIADL RRI+R\n",
      "Sbjct  61   VQKVLKETKAKKVNLIGHSQGPLACRYVAAKHAKNIASVTSINGVNHGSEIADLVRRIVR  120\n",
      "\n",
      "Query  141  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  200\n",
      "            KDS+PEYI + V+ A GTIISTFSGHR +PQDA+AALE+LTTE VTEFN KYPQ LP   \n",
      "Sbjct  121  KDSVPEYIADAVMKAIGTIISTFSGHRSNPQDAVAALEALTTENVTEFNKKYPQGLPAIR  180\n",
      "\n",
      "Query  201  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  260\n",
      "            GGEG E+VNGV+YY FGSYIQGLI GEKGNLLDPTHAAMRVL+ FFTE +NDGLVGR+SM\n",
      "Sbjct  181  GGEGKEVVNGVYYYSFGSYIQGLIVGEKGNLLDPTHAAMRVLSAFFTEHENDGLVGRTSM  240\n",
      "\n",
      "Query  261  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            RLGKLIKDDYA+DH+DMVNQVAGLVG  EDIVAIYT HA +LASK+L\n",
      "Sbjct  241  RLGKLIKDDYAEDHLDMVNQVAGLVGRGEDIVAIYTNHANFLASKKL  287\n",
      "\n",
      "\n",
      ">WP_072062608.1 alpha/beta hydrolase [Proteus vulgaris]\n",
      " CRL58988.1 Lactonizing lipase precursor [Proteus vulgaris]\n",
      "Length=287\n",
      "\n",
      " Score = 480 bits (1236),  Expect = 7e-173, Method: Compositional matrix adjust.\n",
      " Identities = 221/287 (77%), Positives = 258/287 (90%), Gaps = 0/287 (0%)\n",
      "\n",
      "Query  21   MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  80\n",
      "            MST YPIVLVHGL+GF+++VG+PYFYG+ADAL ++GH+VFTASLSAFN+NEVRG+QLW+F\n",
      "Sbjct  1    MSTTYPIVLVHGLSGFDDVVGYPYFYGVADALEKNGHKVFTASLSAFNANEVRGEQLWEF  60\n",
      "\n",
      "Query  81   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  140\n",
      "            VQ +L+ET+AKKVN IGHSQGPLACRYVAA +  S+ASVTSINGVNHGSEIADL R I+R\n",
      "Sbjct  61   VQKVLKETKAKKVNLIGHSQGPLACRYVAAKHAKSIASVTSINGVNHGSEIADLVRHIVR  120\n",
      "\n",
      "Query  141  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  200\n",
      "            KDS+PEYI + V+ +  TIISTFSG+R +PQDA+ ALE+LTTE VTEFN KYPQ LP   \n",
      "Sbjct  121  KDSVPEYIADAVMKSIVTIISTFSGNRSNPQDAVVALEALTTENVTEFNKKYPQGLPAVR  180\n",
      "\n",
      "Query  201  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  260\n",
      "            GGEG E+VNGV+YY FGSYIQGLI GEKGNLLDPTHAAMR+LN FFTE++NDGLVGR+SM\n",
      "Sbjct  181  GGEGKEVVNGVYYYSFGSYIQGLIVGEKGNLLDPTHAAMRILNAFFTERENDGLVGRTSM  240\n",
      "\n",
      "Query  261  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            RLGKLIKDDYA+DH+DMVNQVAGLVG  EDIVAIYT HA++LASK+L\n",
      "Sbjct  241  RLGKLIKDDYAEDHLDMVNQVAGLVGRGEDIVAIYTNHAEFLASKRL  287\n",
      "\n",
      "\n",
      ">WP_064719666.1 alpha/beta hydrolase [Proteus hauseri]\n",
      " OAT47156.1 lipase [Proteus hauseri ATCC 700826]\n",
      "Length=287\n",
      "\n",
      " Score = 470 bits (1209),  Expect = 1e-168, Method: Compositional matrix adjust.\n",
      " Identities = 217/287 (76%), Positives = 255/287 (89%), Gaps = 0/287 (0%)\n",
      "\n",
      "Query  21   MSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQF  80\n",
      "            MST+YPI+LVHGL+GF++IVG+PYFYGIADAL+++GHQVFTASLSAFNSNE+RG+QLW+F\n",
      "Sbjct  1    MSTQYPIILVHGLSGFDDIVGYPYFYGIADALKKEGHQVFTASLSAFNSNEIRGEQLWEF  60\n",
      "\n",
      "Query  81   VQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMR  140\n",
      "            VQ +L+ET+AKKVN IGHSQGPLACRYVAA + + +ASVTS+NGVNHGSEIADL RRI+R\n",
      "Sbjct  61   VQKILKETKAKKVNLIGHSQGPLACRYVAAMHSNKIASVTSVNGVNHGSEIADLVRRIVR  120\n",
      "\n",
      "Query  141  KDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPKTP  200\n",
      "            KDS+PEYI + V+ A G IIS FSG++G+PQ+AIAALE+LTTE V+EFN KYPQ LPK  \n",
      "Sbjct  121  KDSVPEYIADAVMKAIGVIISVFSGNKGNPQNAIAALEALTTESVSEFNKKYPQGLPKKW  180\n",
      "\n",
      "Query  201  GGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRSSM  260\n",
      "            GGEG E+VNGVHYY FGSYIQG I GEK NL D THA MRVL+ FFTEK+NDGLV R SM\n",
      "Sbjct  181  GGEGREVVNGVHYYSFGSYIQGRIIGEKDNLRDLTHAPMRVLSAFFTEKENDGLVSRKSM  240\n",
      "\n",
      "Query  261  RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            RLGKLIKDDY +DH+DMVNQVAGLVG  EDI+AIYT HAK+LA+K+L\n",
      "Sbjct  241  RLGKLIKDDYPEDHLDMVNQVAGLVGRKEDIIAIYTNHAKFLANKKL  287\n",
      "\n",
      "\n",
      ">EEG83604.1 triacylglycerol lipase family protein [Proteus penneri ATCC 35198]\n",
      "Length=169\n",
      "\n",
      " Score = 286 bits (731),  Expect = 8e-98, Method: Compositional matrix adjust.\n",
      " Identities = 131/169 (78%), Positives = 150/169 (89%), Gaps = 0/169 (0%)\n",
      "\n",
      "Query  139  MRKDSIPEYIVEKVLNAFGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQALPK  198\n",
      "            MRKDS+PEYI + V+ A GTIISTFSG+R +PQDA+AALE+LTTE VTEFN KYPQ LP \n",
      "Sbjct  1    MRKDSVPEYIADAVMKAIGTIISTFSGNRSNPQDAVAALEALTTENVTEFNKKYPQGLPA  60\n",
      "\n",
      "Query  199  TPGGEGDEIVNGVHYYCFGSYIQGLIAGEKGNLLDPTHAAMRVLNTFFTEKQNDGLVGRS  258\n",
      "              GGEG E+VNGV+YY FGSYIQG I GEKGNLLDPTHAAMR+LN FFTE++NDGLVGR+\n",
      "Sbjct  61   VRGGEGKEVVNGVYYYSFGSYIQGSIVGEKGNLLDPTHAAMRILNAFFTERENDGLVGRT  120\n",
      "\n",
      "Query  259  SMRLGKLIKDDYAQDHIDMVNQVAGLVGYNEDIVAIYTQHAKYLASKQL  307\n",
      "            SMRLGKL+KDDYA+DH+DMVNQVAGLVG  EDIVAIYT HA++LASK+L\n",
      "Sbjct  121  SMRLGKLVKDDYAEDHLDMVNQVAGLVGRGEDIVAIYTNHAEFLASKRL  169\n",
      "\n",
      "\n",
      ">EEG83605.1 hypothetical protein PROPEN_04374 [Proteus penneri ATCC 35198]\n",
      "Length=81\n",
      "\n",
      " Score = 128 bits (322),  Expect = 2e-37, Method: Composition-based stats.\n",
      " Identities = 56/76 (74%), Positives = 71/76 (93%), Gaps = 0/76 (0%)\n",
      "\n",
      "Query  20  HMSTKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFNSNEVRGKQLWQ  79\n",
      "           +MST YPIVLVHGL+GF+++VG+PYFYGIADAL ++GH+VFTASLSAFN+NEVRG+QLW+\n",
      "Sbjct  3   NMSTTYPIVLVHGLSGFDDVVGYPYFYGIADALEKNGHKVFTASLSAFNANEVRGEQLWE  62\n",
      "\n",
      "Query  80  FVQTLLQETQAKKVNF  95\n",
      "           FVQ +L+ET+AKK  F\n",
      "Sbjct  63  FVQKVLKETKAKKSKF  78\n",
      "\n",
      "\n",
      ">WP_014837821.1 MULTISPECIES: acyl-CoA esterase [Enterobacterales]\n",
      " OFU82673.1 acyl-CoA esterase [Proteus sp. HMSC10D02]\n",
      "Length=257\n",
      "\n",
      " Score = 36.6 bits (83),  Expect = 0.021, Method: Compositional matrix adjust.\n",
      " Identities = 38/121 (31%), Positives = 59/121 (49%), Gaps = 13/121 (11%)\n",
      "\n",
      "Query  26   PIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTAS--LSAFNSNEVRGKQLWQFVQT  83\n",
      "            PIVLVHGL G  + +G      + D   +D  QV   +  LS   S ++    + Q +  \n",
      "Sbjct  18   PIVLVHGLFGSLDNLGILARDLVLD---RDIIQVDMRNHGLSP-RSPDMTYPAMAQDLLD  73\n",
      "\n",
      "Query  84   LLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSINGVNHGSEIADLYRRIMRKDS  143\n",
      "            +L E+Q +K  FIGHS G  A   ++A  P+ +A + +I       +IA +   + R D \n",
      "Sbjct  74   ILNESQIEKATFIGHSMGGKAVMALSALAPERIAGLVAI-------DIAPVDYHVRRHDE  126\n",
      "\n",
      "Query  144  I  144\n",
      "            I\n",
      "Sbjct  127  I  127\n",
      "\n",
      "\n",
      ">WP_036938484.1 acyl-CoA esterase [Proteus vulgaris]\n",
      " KGA59822.1 PGAP1-like family protein [Proteus vulgaris]\n",
      "Length=260\n",
      "\n",
      " Score = 35.8 bits (81),  Expect = 0.040, Method: Compositional matrix adjust.\n",
      " Identities = 31/105 (30%), Positives = 47/105 (45%), Gaps = 8/105 (8%)\n",
      "\n",
      "Query  22   STKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLS---AFNSNEVRGKQLW  78\n",
      "            ++  PIVL+HGL G    +G      I   LRQD   +     +   + +S  +  + + \n",
      "Sbjct  19   ASDLPIVLIHGLFGDLNNLG-----AIGRNLRQDNTVIQIDVRNHGHSPHSESMNYQDMA  73\n",
      "\n",
      "Query  79   QFVQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSIN  123\n",
      "            Q V TLL   Q  K   IGHS G      + A  P+ +A +  I+\n",
      "Sbjct  74   QDVLTLLDSLQISKAIIIGHSMGGKIAMAMTALAPERIARIVVID  118\n",
      "\n",
      "\n",
      ">WP_072070221.1 alpha/beta hydrolase [Proteus vulgaris]\n",
      " CRL64518.1 Esterase YbfF [Proteus vulgaris]\n",
      "Length=260\n",
      "\n",
      " Score = 34.3 bits (77),  Expect = 0.12, Method: Compositional matrix adjust.\n",
      " Identities = 31/106 (29%), Positives = 47/106 (44%), Gaps = 10/106 (9%)\n",
      "\n",
      "Query  22   STKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLS----AFNSNEVRGKQL  77\n",
      "            ++  PIVL+HGL G    +G      +   LRQD H V    +     + +S  +  + +\n",
      "Sbjct  19   ASNLPIVLIHGLFGDLNNLGV-----LGRDLRQD-HTVIQIDVRNHGHSPHSESMNYQDM  72\n",
      "\n",
      "Query  78   WQFVQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSIN  123\n",
      "             Q V TLL      K   IGHS G      + A  P+ +A +  I+\n",
      "Sbjct  73   AQDVLTLLDSLHISKAIIIGHSMGGKIAMAMTALAPERIARIVVID  118\n",
      "\n",
      "\n",
      ">WP_064719330.1 acyl-CoA esterase [Proteus hauseri]\n",
      " OAT47838.1 esterase [Proteus hauseri ATCC 700826]\n",
      "Length=263\n",
      "\n",
      " Score = 33.5 bits (75),  Expect = 0.24, Method: Compositional matrix adjust.\n",
      " Identities = 33/106 (31%), Positives = 43/106 (41%), Gaps = 14/106 (13%)\n",
      "\n",
      "Query  24   KYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFN------SNEVRGKQL  77\n",
      "            + PIVL+HGL G    +G      +A  LRQD        +   N      SN +    +\n",
      "Sbjct  24   QVPIVLIHGLFGDLNNLGV-----LARDLRQDN---IVIQIDVRNHGHSPHSNSMNYLDM  75\n",
      "\n",
      "Query  78   WQFVQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSIN  123\n",
      "             Q V TLL      K   IGHS G      + A  PD +  V  I+\n",
      "Sbjct  76   AQDVITLLDSLNIAKAIVIGHSMGGKIAMAMTALAPDRLDKVVVID  121\n",
      "\n",
      "\n",
      ">WP_075671659.1 alpha/beta hydrolase [Proteus sp. H24]\n",
      "Length=260\n",
      "\n",
      " Score = 33.5 bits (75),  Expect = 0.25, Method: Compositional matrix adjust.\n",
      " Identities = 32/113 (28%), Positives = 46/113 (41%), Gaps = 24/113 (21%)\n",
      "\n",
      "Query  22   STKYPIVLVHGLAGFNEIVGFPYFYGIADALRQD-----------GHQVFTASLSAFNSN  70\n",
      "            ++  PIVL+HGL G    +G      +  ALRQD           GH     S++     \n",
      "Sbjct  19   ASNLPIVLIHGLFGDLNNLGV-----LGRALRQDNTVIQIDVRNHGHSPHCESMNY----  69\n",
      "\n",
      "Query  71   EVRGKQLWQFVQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSIN  123\n",
      "                + + Q V TLL      K   IGHS G      + A  P  +A++  I+\n",
      "Sbjct  70   ----QDMAQDVLTLLDSLNVSKAIIIGHSMGGKIAMAMTALAPQRIANIVVID  118\n",
      "\n",
      "\n",
      ">WP_072062930.1 alpha/beta hydrolase [Proteus vulgaris]\n",
      " CRL59751.1 Esterase YbfF [Proteus vulgaris]\n",
      "Length=262\n",
      "\n",
      " Score = 33.1 bits (74),  Expect = 0.37, Method: Compositional matrix adjust.\n",
      " Identities = 29/105 (28%), Positives = 48/105 (46%), Gaps = 8/105 (8%)\n",
      "\n",
      "Query  22   STKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLS---AFNSNEVRGKQLW  78\n",
      "            ++  PIVL+HGL G    +G      ++ ALRQD   +     +   + +S  +  + + \n",
      "Sbjct  21   ASNLPIVLIHGLFGDLNNLGV-----LSRALRQDNTVIQIDVRNHGHSPHSESMNYQDMA  75\n",
      "\n",
      "Query  79   QFVQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSIN  123\n",
      "            Q V  LL      K   IGHS G      + A  P+ +A++  I+\n",
      "Sbjct  76   QDVLALLDNLAIPKAIIIGHSMGGKIAMAMTALAPERIANIVVID  120\n",
      "\n",
      "\n",
      ">EEG84125.1 PGAP1-like protein [Proteus penneri ATCC 35198]\n",
      "Length=233\n",
      "\n",
      " Score = 32.3 bits (72),  Expect = 0.50, Method: Compositional matrix adjust.\n",
      " Identities = 30/108 (28%), Positives = 47/108 (44%), Gaps = 14/108 (13%)\n",
      "\n",
      "Query  22   STKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGHQVFTASLSAFN------SNEVRGK  75\n",
      "            ++  PIVL+HGL G    +G      ++ ALRQD   +    +   N      S  +  +\n",
      "Sbjct  36   ASNLPIVLIHGLFGDLNNLGV-----LSRALRQDNTVI---QIDVRNHGHSPHSESMNYQ  87\n",
      "\n",
      "Query  76   QLWQFVQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSIN  123\n",
      "             + Q V  LL      K   IGHS G      + A  P+ +A++  I+\n",
      "Sbjct  88   DMAQDVLALLDNLAIPKAIIIGHSMGGKIAMAMTALAPERIANIVVID  135\n",
      "\n",
      "\n",
      ">WP_023581172.1 hydrolase [Proteus hauseri]\n",
      " EST59347.1 hydrolase [Proteus hauseri ZMd44]\n",
      "Length=264\n",
      "\n",
      " Score = 32.3 bits (72),  Expect = 0.60, Method: Compositional matrix adjust.\n",
      " Identities = 32/105 (30%), Positives = 46/105 (44%), Gaps = 8/105 (8%)\n",
      "\n",
      "Query  22   STKYPIVLVHGLAGFNEIVGFPYFYGIADALRQDGH--QVFTASLSAFNSNEVRGKQ-LW  78\n",
      "            ++  PIVL+HGL G    +G      +  ALRQD    Q+   +      NE    Q + \n",
      "Sbjct  19   ASHLPIVLIHGLFGDLNNLGV-----LGRALRQDNTVIQIDVRNHGHSPHNESMNYQDMA  73\n",
      "\n",
      "Query  79   QFVQTLLQETQAKKVNFIGHSQGPLACRYVAANYPDSVASVTSIN  123\n",
      "            Q V  LL      K   IGHS G      + A  P+ +A++  I+\n",
      "Sbjct  74   QDVLALLDSLAISKAIIIGHSMGGKIAMAMTALAPERIANIVVID  118\n",
      "\n",
      "\n",
      ">AMR60768.1 replicase (plasmid) [Proteus mirabilis]\n",
      " AMR60774.1 replicase (plasmid) [Proteus mirabilis]\n",
      "Length=272\n",
      "\n",
      " Score = 30.4 bits (67),  Expect = 2.5, Method: Compositional matrix adjust.\n",
      " Identities = 21/75 (28%), Positives = 33/75 (44%), Gaps = 8/75 (11%)\n",
      "\n",
      "Query  143  SIPEYIVEKVLNA-------FGTIISTFSGHRGDPQDAIAALESLTTEQVTEFNNKYPQA  195\n",
      "            ++P +I +  LN          T +S     RG PQ  +A ++   TEQ+ E +  Y   \n",
      "Sbjct  71   AMPNWICQNPLNGHAHYGYMLKTPVSRTLKARGTPQRYLARIQQAMTEQL-EADAGYAHF  129\n",
      "\n",
      "Query  196  LPKTPGGEGDEIVNG  210\n",
      "            L KTP  +    + G\n",
      "Sbjct  130  LTKTPAHDKHRTIWG  144\n",
      "\n",
      "\n",
      ">WP_072196818.1 hypothetical protein [Proteus mirabilis]\n",
      "Length=63\n",
      "\n",
      " Score = 26.6 bits (57),  Expect = 9.3, Method: Composition-based stats.\n",
      " Identities = 14/44 (32%), Positives = 22/44 (50%), Gaps = 5/44 (11%)\n",
      "\n",
      "Query  253  GLVGRSSM-----RLGKLIKDDYAQDHIDMVNQVAGLVGYNEDI  291\n",
      "            GL+G S++     R+ KL+   Y QDH +    +   + YN  I\n",
      "Sbjct  20   GLIGNSTVILTANRINKLLLSQYNQDHPETARFLPIFIAYNYSI  63\n",
      "\n",
      "\n",
      "  Database: All non-redundant GenBank CDS translations+PDB+SwissProt+PIR+PRF\n",
      "excluding environmental samples from WGS projects\n",
      "    Posted date:  Mar 6, 2017  3:34 AM\n",
      "  Number of letters in database: 25,458,496\n",
      "  Number of sequences in database:  76,021\n",
      "\n",
      "Lambda      K        H\n",
      "   0.318    0.135    0.398 \n",
      "Gapped\n",
      "Lambda      K        H\n",
      "   0.267   0.0410    0.140 \n",
      "Matrix: BLOSUM62\n",
      "Gap Penalties: Existence: 11, Extension: 1\n",
      "Number of Sequences: 76021\n",
      "Number of Hits to DB: 160850\n",
      "Number of extensions: 6849\n",
      "Number of successful extensions: 21\n",
      "Number of sequences better than 100: 4\n",
      "Number of HSP's better than 100 without gapping: 0\n",
      "Number of HSP's gapped: 21\n",
      "Number of HSP's successfully gapped: 4\n",
      "Length of query: 307\n",
      "Length of database: 25458496\n",
      "Length adjustment: 103\n",
      "Effective length of query: 204\n",
      "Effective length of database: 17628333\n",
      "Effective search space: 3596179932\n",
      "Effective search space used: 3596179932\n",
      "T: 11\n",
      "A: 40\n",
      "X1: 16 (7.3 bits)\n",
      "X2: 38 (14.6 bits)\n",
      "X3: 64 (24.7 bits)\n",
      "S1: 41 (20.4 bits)\n",
      "S2: 54 (25.4 bits)\n",
      "ka-blk-alpha gapped: 1.9\n",
      "ka-blk-alpha ungapped: 0.7916\n",
      "ka-blk-alpha_v gapped: 42.6028\n",
      "ka-blk-alpha_v ungapped: 4.96466\n",
      "ka-blk-sigma gapped: 43.6362\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Get human-readable results\n",
    "output = result_handle.read()\n",
    "\n",
    "# Show result\n",
    "print(output)\n",
    "\n",
    "# Save output to file\n",
    "outfilename = os.path.join(\"output\", \"wildtype_remote_blastp_nr.txt\")\n",
    "with open(outfilename, 'w') as outfh:\n",
    "    outfh.write(output)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Option 2: in the browser**\n",
    "\n",
    "* Upload the `output/wildtype_aa.fasta` sequence\n",
    "* Select `nr`\n",
    "* Put the term `proteus[ORGN]` into the `Entrez query` field\n",
    "* Submit\n",
    "\n",
    "The results will come back graphically in a couple of minutes\n",
    "\n",
    "* The search returns 22 matches\n",
    "  * Click on `Download -> Download alignment -> Text` to download the plain text results\n",
    "  * Click on `Download -> Download alignment -> Table(csv)` to download tabular results\n",
    "* Click on `Multiple alignment` to see an alignment\n",
    "  * Click on `Download -> Download alignment -> FASTA plus gaps` (this will go in the Downloads directory)\n",
    "* Click on `MSA viewer` to see the MSA\n",
    "* Click on `Tools -> Coloring -> Rasmol amino acid colors` to see differences highlighted.\n",
    "\n",
    "<p></p>\n",
    "* **Ask the learners to move the downloaded files to the `output` directory**\n",
    "* **NOTE: the source structures are two of the hits recovered in the alignment. You can point this out if you like.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 02b - UniProt\n",
    "\n",
    "We can query UniProt in the browser or from the notebook. We have a match from the local reference genome that is 100% identity over the entire sequence, so we can be confident that the functional annotation is for the wildtype protein."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert-success\">\n",
    "The `PMI0999` sequence obtained in the `BLAST` search is identifcal to the wildtype, so this accession was used as a query against `UniProt` to identify database cross-links and functional information.\n",
    "</div>\n",
    "\n",
    "**Option 1: in the browser**\n",
    "\n",
    "* **Ask the learners to enter `PMI0999` as a query at the `UniProt` home page**\n",
    "* **Ask the learners to obtain the mappings of this entry to `GenBank`, `KEGG` (and `KO`), `PFam`, `Ensembl`, `STRING`, and `PDB`, if available**\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Option 2: in the notebook**\n",
    "\n",
    "* **Ask the learners to use `PMI0999` as a query for `UniProt` to identify the entry accession**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Entry\tEntry name\tStatus\tProtein names\tGene names\tOrganism\tLength\n",
      "B4EVM3\tB4EVM3_PROMH\tunreviewed\tPutative lipase (EC 3.1.1.3)\tPMI0999\tProteus mirabilis (strain HI4320)\t287\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Make a link to the UniProt webservice\n",
    "service = UniProt()\n",
    "\n",
    "# Make a new query string, and run a remote search at UniProt\n",
    "new_query = \"PMI0999\"\n",
    "new_result = service.search(new_query)\n",
    "\n",
    "# Inspect the result\n",
    "print(new_result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **Ask the learners to obtain the `pandas` dataframe result for this accession**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry name</th>\n",
       "      <th>Gene names</th>\n",
       "      <th>Gene names  (primary )</th>\n",
       "      <th>Gene names  (synonym )</th>\n",
       "      <th>Gene names  (ordered locus )</th>\n",
       "      <th>Gene names  (ORF )</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Organism ID</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>...</th>\n",
       "      <th>Miscellaneous [CC]</th>\n",
       "      <th>Keywords</th>\n",
       "      <th>Protein existence</th>\n",
       "      <th>Status</th>\n",
       "      <th>Sequence annotation (Features)</th>\n",
       "      <th>Protein families</th>\n",
       "      <th>Version</th>\n",
       "      <th>Comments</th>\n",
       "      <th>Cross-reference</th>\n",
       "      <th>Pathway.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>B4EVM3</td>\n",
       "      <td>B4EVM3_PROMH</td>\n",
       "      <td>[PMI0999]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PMI0999</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Proteus mirabilis (strain HI4320)</td>\n",
       "      <td>529507</td>\n",
       "      <td>Putative lipase (EC 3.1.1.3)</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[3D-structure, Calcium, Complete proteome, Hyd...</td>\n",
       "      <td>Evidence at protein level</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[]</td>\n",
       "      <td>53</td>\n",
       "      <td>[]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 100 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Entry    Entry name Gene names  Gene names  (primary )  \\\n",
       "0  B4EVM3  B4EVM3_PROMH  [PMI0999]                     NaN   \n",
       "\n",
       "   Gene names  (synonym ) Gene names  (ordered locus )  Gene names  (ORF )  \\\n",
       "0                     NaN                      PMI0999                 NaN   \n",
       "\n",
       "                            Organism  Organism ID  \\\n",
       "0  Proteus mirabilis (strain HI4320)       529507   \n",
       "\n",
       "                  Protein names    ...     Miscellaneous [CC]  \\\n",
       "0  Putative lipase (EC 3.1.1.3)    ...                    NaN   \n",
       "\n",
       "                                            Keywords  \\\n",
       "0  [3D-structure, Calcium, Complete proteome, Hyd...   \n",
       "\n",
       "           Protein existence      Status Sequence annotation (Features)  \\\n",
       "0  Evidence at protein level  unreviewed                            NaN   \n",
       "\n",
       "   Protein families Version  Comments  Cross-reference  Pathway.1  \n",
       "0                []      53        []              NaN        NaN  \n",
       "\n",
       "[1 rows x 100 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get a dataframe for a single accession ID\n",
    "df = service.get_df('B4EVM3')\n",
    "\n",
    "# View the dataframe\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KEGG: defaultdict(<class 'list'>, {'B4EVM3': ['pmr:PMI0999']})\n",
      "KEGG KO: defaultdict(<class 'list'>, {'B4EVM3': ['K01046']})\n",
      "Ensembl Protein: defaultdict(<class 'list'>, {'B4EVM3': ['CAR42190']})\n",
      "Ensembl Transcript: defaultdict(<class 'list'>, {'B4EVM3': ['CAR42190']})\n",
      "STRING: defaultdict(<class 'list'>, {'B4EVM3': ['529507.PMI0999']})\n",
      "GENBANK: defaultdict(<class 'list'>, {'B4EVM3': ['AM942759']})\n",
      "PDB: defaultdict(<class 'list'>, {'B4EVM3': ['3W9U', '4GW3']})\n",
      "OrthoDB: defaultdict(<class 'list'>, {'B4EVM3': ['POG091H09SQ']})\n"
     ]
    }
   ],
   "source": [
    "# Use the mapping() method of the bioservices service to obtain database cross-references\n",
    "# See also http://www.uniprot.org/help/programmatic_access\n",
    "entry = \"B4EVM3\"\n",
    "kegg_links = service.mapping(\"ACC\", \"KEGG_ID\", entry)\n",
    "kegg_ko_links = service.mapping(\"ACC\", \"KO_ID\", entry)\n",
    "ensembl_protein_links = service.mapping(\"ACC\", \"ENSEMBLGENOME_PRO_ID\", entry)\n",
    "ensembl_transcript_links = service.mapping(\"ACC\", \"ENSEMBLGENOME_TRS_ID\", entry)\n",
    "string_links =service.mapping(\"ACC\", \"STRING_ID\", entry)\n",
    "genbank_links = service.mapping(\"ACC\", \"EMBL_ID\", entry)\n",
    "pdb_links = service.mapping(\"ACC\", \"PDB_ID\", entry)\n",
    "orthodb_links = service.mapping(\"ACC\", \"ORTHODB_ID\", entry)\n",
    "\n",
    "# Print links\n",
    "print(\"KEGG: {0}\".format(kegg_links))\n",
    "print(\"KEGG KO: {0}\".format(kegg_ko_links))\n",
    "print(\"Ensembl Protein: {0}\".format(ensembl_protein_links))\n",
    "print(\"Ensembl Transcript: {0}\".format(ensembl_transcript_links))\n",
    "print(\"STRING: {0}\".format(string_links))\n",
    "print(\"GENBANK: {0}\".format(genbank_links))\n",
    "print(\"PDB: {0}\".format(pdb_links))\n",
    "print(\"OrthoDB: {0}\".format(orthodb_links))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert-success\">\n",
    "The `PMI0999` accession corresponds to `UniProt` entry `B4EVM3`, an unreviewed putative lipase from *Proteus mirabilis*. The database crossreferences are:\n",
    "<ul>\n",
    "  <li> `GenBank`: CAR42190.1; AM942759\n",
    "  <li> `KEGG`: pmr:PMI0999\n",
    "  <li> `KEGG KO`: pmr:K01046  \n",
    "  <li> `Ensembl Protein`: CAR42190\n",
    "  <li> `Ensembl Transcript`: CAR42190\n",
    "  <li> `PFam`: PF00561 (Abhydrolase_1)\n",
    "  <li> `Ensembl`: CAR42190; PMI0999\n",
    "  <li> `STRING`: 529507.PMI0999\n",
    "  <li> `PDB`: 2W9U, 4GW3\n",
    "  <li> `OrthoDB`: POG091H09SQ\n",
    "</ul>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (IBioIC)",
   "language": "python",
   "name": "python3_ibioic_course"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
